## Capstone Project Data Exploration, Preparation, EDA Notebook

### Question
### Is homeownership still attainable in the US?
### What economic factors (median income, educational attainment, home prices) influence middle-income American households’ ability to buy a home, and what US cities have become the least and most attainable and populated?

### We will be exploring data from 2010, 2015, and 2024, to evaluate trends spanning the last 15 years


#### Datasets used:
#### From Zillow:
##### Zillow Home Value Index (ZHVI). URL: https://www.zillow.com/research/data/

#### From Census.gov:
##### Median Household Income Past 12 months 2024. URL https://data.census.gov/table/ACSDT1Y2024.B19013?t=Income+(Households,+Families,+Individuals):Income+and+Earnings&g=010XX00US$31000M1&y=2024

##### Median Household Income Past 12 months 2015. URL https://data.census.gov/table/ACSDT5Y2015.B19013?t=Income+(Households,+Families,+Individuals):Income+and+Earnings&g=010XX00US$31000M1&y=2015

##### Median Household Income Past 12 months 2010. URL https://data.census.gov/table/ACSDT5Y2010.B19013?t=Income+(Households,+Families,+Individuals):Income+and+Earnings&g=010XX00US$31000M1&y=2010

##### Total Population 2024 (Metro). URL https://data.census.gov/table/ACSDT5Y2024.B01003?t=Counts,+Estimates,+and+Projections:Population+Total&g=010XX00US$31000M1&y=2024

##### Total Population 2015 (Metro). URL https://data.census.gov/table/ACSDT5Y2015.B01003?t=Counts,+Estimates,+and+Projections:Population+Total&g=010XX00US$31000M1&y=2015

##### Total Population 2010 (Metro). URL https://data.census.gov/table/ACSDT5Y2010.B01003?t=Counts,+Estimates,+and+Projections:Population+Total&g=010XX00US$31000M1&y=2010

##### Educational Attainment for POP over 25 2024. URL https://data.census.gov/table/ACSDT5Y2024.B15003?t=Educational+Attainment&g=010XX00US$31000M1&y=2024

##### Educational Attainment for POP over 25 2015. URL https://data.census.gov/table/ACSDT5Y2015.B15003?t=Educational+Attainment&g=010XX00US$31000M1&y=2015

##### Educational Attainment for POP over 25 2010. URL https://data.census.gov/table/ACSDT1Y2010.B15003?t=Educational+Attainment&g=010XX00US$31000M1&y=2010



In [2]:
import pandas as pd
import numpy as np

### Exploring Zillow Dataset
#### and some EDA

In [3]:
zillow = pd.read_csv('Data Raw/Zillow_home_values_30.csv')

In [4]:
zillow

,RegionID,SizeRank,RegionName,RegionType,StateName,2000-01-31,2000-02-29,2000-03-31,2000-04-30,2000-05-31,...,2025-07-31,2025-08-31,2025-09-30,2025-10-31,2025-11-30,2025-12-31,2026-01-31,2026-02-28,2026-03-31,2026-04-30
0,102001,0,United States,country,NaN,123427.049538,123644.197286,123913.492467,124491.077769,125155.809204,...,364209.427056,363979.753025,364223.486665,364661.623597,365343.536055,366141.311822,366902.442255,367622.830024,368117.684407,368197.746706
1,394913,1,"New York, NY",msa,NY,217908.609033,218834.597673,219769.162637,221662.952885,223624.038269,...,694306.523004,695095.503604,696494.286361,698826.820996,702164.818651,705580.479548,708640.135014,712208.505439,715991.562187,718396.823077
2,753899,2,"Los Angeles, CA",msa,CA,225552.897453,226392.240410,227510.237225,229734.675324,232167.513239,...,957897.593304,956807.033618,957882.766298,960342.385171,963719.386447,967750.414641,970325.833108,971579.813792,971239.927443,969588.759213
3,394463,3,"Chicago, IL",msa,IL,155258.141365,155401.865176,155675.589704,156356.548308,157175.659600,...,337066.458681,338073.991308,339422.785924,340743.001060,342308.924510,344061.772545,345700.958158,347468.224755,348890.551227,349901.365159
4,394514,4,"Dallas, TX",msa,TX,130734.453703,130792.683285,130859.624151,131033.462549,131262.663716,...,373108.019068,371712.835496,371050.042794,370692.979211,370422.910339,370199.905274,369755.843945,369166.664518,368086.223059,366742.535921
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
890,753929,935,"Zapata, TX",msa,TX,NaN,NaN,NaN,NaN,NaN,...,126101.825750,125642.649749,124871.905124,123885.184463,123092.686822,123053.236536,123810.858888,124612.138446,125635.764563,125226.746254
891,394743,936,"Ketchikan, AK",msa,AK,NaN,NaN,NaN,NaN,NaN,...,377893.253426,382116.813311,385628.591267,388984.295165,391409.920696,393482.128114,394719.215011,395439.487256,396972.690796,399056.055088
892,753874,937,"Craig, CO",msa,CO,99513.105944,99767.567915,100241.309446,100915.961146,101699.952603,...,301229.439262,302250.397134,303517.714708,303810.683032,303362.905348,302857.646971,303007.310021,303712.635345,304618.446587,305274.568594
893,395188,938,"Vernon, TX",msa,TX,NaN,NaN,NaN,NaN,NaN,...,102205.659403,101922.178860,101532.485847,101318.898110,101839.845936,102830.379874,103088.027459,103385.170785,103472.279661,103952.124110


In [5]:
zillow.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 895 entries, 0 to 894
Columns: 321 entries, RegionID to 2026-04-30
dtypes: float64(316), int64(2), object(3)
memory usage: 2.2+ MB


In [6]:
pd.melt(
    zillow,
    id_vars=['RegionName','StateName'],
    var_name='Date',
    value_name='HomeValue'
)

,RegionName,StateName,Date,HomeValue
0,United States,NaN,RegionID,102001
1,"New York, NY",NY,RegionID,394913
2,"Los Angeles, CA",CA,RegionID,753899
3,"Chicago, IL",IL,RegionID,394463
4,"Dallas, TX",TX,RegionID,394514
...,...,...,...,...
285500,"Zapata, TX",TX,2026-04-30,125226.746254
285501,"Ketchikan, AK",AK,2026-04-30,399056.055088
285502,"Craig, CO",CO,2026-04-30,305274.568594
285503,"Vernon, TX",TX,2026-04-30,103952.12411


In [7]:
zillow.isna().sum()

RegionID      0
SizeRank      0
RegionName    0
RegionType    0
StateName     1
             ..
2025-12-31    0
2026-01-31    0
2026-02-28    0
2026-03-31    0
2026-04-30    0
Length: 321, dtype: int64

In [8]:
zillow['RegionName'].nunique()

895

In [9]:
zillow.columns.tolist()

['RegionID',
 'SizeRank',
 'RegionName',
 'RegionType',
 'StateName',
 '2000-01-31',
 '2000-02-29',
 '2000-03-31',
 '2000-04-30',
 '2000-05-31',
 '2000-06-30',
 '2000-07-31',
 '2000-08-31',
 '2000-09-30',
 '2000-10-31',
 '2000-11-30',
 '2000-12-31',
 '2001-01-31',
 '2001-02-28',
 '2001-03-31',
 '2001-04-30',
 '2001-05-31',
 '2001-06-30',
 '2001-07-31',
 '2001-08-31',
 '2001-09-30',
 '2001-10-31',
 '2001-11-30',
 '2001-12-31',
 '2002-01-31',
 '2002-02-28',
 '2002-03-31',
 '2002-04-30',
 '2002-05-31',
 '2002-06-30',
 '2002-07-31',
 '2002-08-31',
 '2002-09-30',
 '2002-10-31',
 '2002-11-30',
 '2002-12-31',
 '2003-01-31',
 '2003-02-28',
 '2003-03-31',
 '2003-04-30',
 '2003-05-31',
 '2003-06-30',
 '2003-07-31',
 '2003-08-31',
 '2003-09-30',
 '2003-10-31',
 '2003-11-30',
 '2003-12-31',
 '2004-01-31',
 '2004-02-29',
 '2004-03-31',
 '2004-04-30',
 '2004-05-31',
 '2004-06-30',
 '2004-07-31',
 '2004-08-31',
 '2004-09-30',
 '2004-10-31',
 '2004-11-30',
 '2004-12-31',
 '2005-01-31',
 '2005-02-28',


In [10]:
zillow.columns[:10]

Index(['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName',
       '2000-01-31', '2000-02-29', '2000-03-31', '2000-04-30', '2000-05-31'],
      dtype='object')

In [11]:
zillow_long = pd.melt(
    zillow,
    id_vars=['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName'],
    var_name='Date',
    value_name='HomeValue'
)

In [12]:
zillow_long['Date'] = pd.to_datetime(zillow_long['Date'])
zillow_long['Year'] = zillow_long['Date'].dt.year

In [13]:
zillow_long.head()

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
0,102001,0,United States,country,NaN,2000-01-31,123427.049538,2000
1,394913,1,"New York, NY",msa,NY,2000-01-31,217908.609033,2000
2,753899,2,"Los Angeles, CA",msa,CA,2000-01-31,225552.897453,2000
3,394463,3,"Chicago, IL",msa,IL,2000-01-31,155258.141365,2000
4,394514,4,"Dallas, TX",msa,TX,2000-01-31,130734.453703,2000


In [14]:
zillow_long.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 282820 entries, 0 to 282819
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   RegionID    282820 non-null  int64         
 1   SizeRank    282820 non-null  int64         
 2   RegionName  282820 non-null  object        
 3   RegionType  282820 non-null  object        
 4   StateName   282504 non-null  object        
 5   Date        282820 non-null  datetime64[ns]
 6   HomeValue   233554 non-null  float64       
 7   Year        282820 non-null  int32         
dtypes: datetime64[ns](1), float64(1), int32(1), int64(2), object(3)
memory usage: 16.2+ MB


In [15]:
zillow_msa = zillow_long[zillow_long['RegionType'] == 'msa'].copy()

In [16]:
zillow_msa['RegionName'].nunique()

894

In [17]:
zillow_msa['Date'].min()

Timestamp('2000-01-31 00:00:00')

In [18]:
zillow_msa['Date'].max()

Timestamp('2026-04-30 00:00:00')

In [19]:
zillow_msa['HomeValue'].isna().sum()

np.int64(49266)

In [20]:
zillow_msa

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
1,394913,1,"New York, NY",msa,NY,2000-01-31,217908.609033,2000
2,753899,2,"Los Angeles, CA",msa,CA,2000-01-31,225552.897453,2000
3,394463,3,"Chicago, IL",msa,IL,2000-01-31,155258.141365,2000
4,394514,4,"Dallas, TX",msa,TX,2000-01-31,130734.453703,2000
5,394692,5,"Houston, TX",msa,TX,2000-01-31,125738.933364,2000
...,...,...,...,...,...,...,...,...
282815,753929,935,"Zapata, TX",msa,TX,2026-04-30,125226.746254,2026
282816,394743,936,"Ketchikan, AK",msa,AK,2026-04-30,399056.055088,2026
282817,753874,937,"Craig, CO",msa,CO,2026-04-30,305274.568594,2026
282818,395188,938,"Vernon, TX",msa,TX,2026-04-30,103952.124110,2026


## Zillow Home Value Index

Source:
Zillow Research

Region Type:
MSA

Coverage:
894 metropolitan areas

Date Range:
January 2000 – April 2026

Observations:
282,504

Missing Home Values:
49,266

Transformations:
- Converted wide format to long format using pd.melt()
- Converted Date to datetime
- Extracted Year
- Filtered to RegionType == 'msa'

In [21]:
zillow_msa[['RegionName','StateName']].drop_duplicates().head(20)

,RegionName,StateName
1,"New York, NY",NY
2,"Los Angeles, CA",CA
3,"Chicago, IL",IL
4,"Dallas, TX",TX
5,"Houston, TX",TX
6,"Washington, DC",VA
7,"Philadelphia, PA",PA
8,"Miami, FL",FL
9,"Atlanta, GA",GA
10,"Boston, MA",MA


In [22]:
##Dataset Inventory
dataset_inventory = pd.DataFrame({
    'Dataset': [
        'Zillow_home_values_30',
        'zillow_long',
        'zillow_msa',
        'Median_income_2010',
        'Median_income_2015',
        'Median_income_2024',
        'Pop_2010',
        'Pop_2015',
        'Pop_2024',
        'median_income2024_tidy',
        'merged_2024'
    ],
    'Source': [
        'Zillow',
        'Zillow',
        'Zillow',
        'Census',
        'Census',
        'Census',
        'Census',
        'Census',
        'Census',
        'Census',
        'Zillow+Census'
    ],
    'Geography': [
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA',
        'MSA'
    ],
    'Years': [
        '2000-2026',
        '2000-2026',
        '2000-2026',
        '2010',
        '2015',
        '2024',
        '2010',
        '2015',
        '2024',
        '2024',
        '2024'
    ],
    'Variables': [
        'HomeValue',
        'HomeValue',
        'HomeValue',
        'Median Income',
        'Median Income',
        'Median Income',
        'Population',
        'Population',
        'Population',
        'Median Income',
        'Home + Income'
        
    ],
    'Status': [
        'Cleaned & Melted',
        'Cleaned & Melted',
        'Cleaned & Melted',
        'Not Reviewed',
        'Not Reviewed',
        'Not Reviewed',
        'Not Reviewed',
        'Not Reviewed',
        'Not Reviewed',
        'Saved',
        'Merged+Saved'
    ],
    'Location': [
        'Data Raw',
        'Data Intermediate',
        'Data Intermediate',
        'Data Raw',
        'Data Raw',
        'Data Raw',
        'Data Raw',
        'Data Raw',
        'Data Raw',
        'Data Intermediate',
        'Data Intermediate'
    ]
})
dataset_inventory

,Dataset,Source,Geography,Years,Variables,Status,Location
0,Zillow_home_values_30,Zillow,MSA,2000-2026,HomeValue,Cleaned & Melted,Data Raw
1,zillow_long,Zillow,MSA,2000-2026,HomeValue,Cleaned & Melted,Data Intermediate
2,zillow_msa,Zillow,MSA,2000-2026,HomeValue,Cleaned & Melted,Data Intermediate
3,Median_income_2010,Census,MSA,2010,Median Income,Not Reviewed,Data Raw
4,Median_income_2015,Census,MSA,2015,Median Income,Not Reviewed,Data Raw
5,Median_income_2024,Census,MSA,2024,Median Income,Not Reviewed,Data Raw
6,Pop_2010,Census,MSA,2010,Population,Not Reviewed,Data Raw
7,Pop_2015,Census,MSA,2015,Population,Not Reviewed,Data Raw
8,Pop_2024,Census,MSA,2024,Population,Not Reviewed,Data Raw
9,median_income2024_tidy,Census,MSA,2024,Median Income,Saved,Data Intermediate


#### Importing Median Income 2024 Dataset

In [23]:
median_income2024 = pd.read_csv('Data Raw/Median_income_2024.csv')

In [24]:
median_income2024

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla, PR Metro Area!!Estimate","Aguadilla, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren, OH Metro Area!!Estimate","Youngstown-Warren, OH Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Median household income in the past 12 months ...,"63,390","±7,471","24,351","±1,426","71,364","±1,942","56,851","±4,592","80,422",...,"75,399","±5,440","81,453","±2,464","57,812","±2,859","81,094","±7,657","66,844","±5,762"


In [25]:
median_income2024.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla, PR Metro Area!!Estimate","Aguadilla, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren, OH Metro Area!!Estimate","Youngstown-Warren, OH Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Median household income in the past 12 months ...,"63,390","±7,471","24,351","±1,426","71,364","±1,942","56,851","±4,592","80,422",...,"75,399","±5,440","81,453","±2,464","57,812","±2,859","81,094","±7,657","66,844","±5,762"


In [26]:
median_income2024.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Columns: 787 entries, Label (Grouping) to Yuma, AZ Metro Area!!Margin of Error
dtypes: object(787)
memory usage: 6.3+ KB


In [27]:
median_income2024.columns

Index(['Label (Grouping)', 'Abilene, TX Metro Area!!Estimate',
       'Abilene, TX Metro Area!!Margin of Error',
       'Aguadilla, PR Metro Area!!Estimate',
       'Aguadilla, PR Metro Area!!Margin of Error',
       'Akron, OH Metro Area!!Estimate',
       'Akron, OH Metro Area!!Margin of Error',
       'Albany, GA Metro Area!!Estimate',
       'Albany, GA Metro Area!!Margin of Error',
       'Albany, OR Metro Area!!Estimate',
       ...
       'Yakima, WA Metro Area!!Estimate',
       'Yakima, WA Metro Area!!Margin of Error',
       'York-Hanover, PA Metro Area!!Estimate',
       'York-Hanover, PA Metro Area!!Margin of Error',
       'Youngstown-Warren, OH Metro Area!!Estimate',
       'Youngstown-Warren, OH Metro Area!!Margin of Error',
       'Yuba City, CA Metro Area!!Estimate',
       'Yuba City, CA Metro Area!!Margin of Error',
       'Yuma, AZ Metro Area!!Estimate',
       'Yuma, AZ Metro Area!!Margin of Error'],
      dtype='object', length=787)

#### Importing Median Income 2010 Dataset

In [28]:
median_income2010 = pd.read_csv('Data Raw/Median_income_2010.csv')

In [29]:
median_income2010.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Estimate","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany-Schenectady-Troy, NY Metro Area!!Estimate",...,"Yauco, PR Metro Area!!Estimate","Yauco, PR Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate","Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Median household income in the past 12 months ...,"42,286","±1,168","13,489",±366,"48,570",±589,"36,828","±1,221","57,781",...,"13,911",±433,"57,494",±817,"41,474",±455,"49,190","±1,309","40,340","±1,162"


In [30]:
median_income2010.columns

Index(['Label (Grouping)', 'Abilene, TX Metro Area!!Estimate',
       'Abilene, TX Metro Area!!Margin of Error',
       'Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Estimate',
       'Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Margin of Error',
       'Akron, OH Metro Area!!Estimate',
       'Akron, OH Metro Area!!Margin of Error',
       'Albany, GA Metro Area!!Estimate',
       'Albany, GA Metro Area!!Margin of Error',
       'Albany-Schenectady-Troy, NY Metro Area!!Estimate',
       ...
       'Yauco, PR Metro Area!!Estimate',
       'Yauco, PR Metro Area!!Margin of Error',
       'York-Hanover, PA Metro Area!!Estimate',
       'York-Hanover, PA Metro Area!!Margin of Error',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error',
       'Yuba City, CA Metro Area!!Estimate',
       'Yuba City, CA Metro Area!!Margin of Error',
       'Yuma, AZ Metro Area!!Estimate',
       'Yuma, AZ Metro Area!!M

In [31]:
median_income2010.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Columns: 749 entries, Label (Grouping) to Yuma, AZ Metro Area!!Margin of Error
dtypes: object(749)
memory usage: 6.0+ KB


#### Importing Median Income 2015 Dataset

In [32]:
median_income2015 = pd.read_csv('Data Raw/Median_income_2015.csv')

In [33]:
median_income2015.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla-Isabela, PR Metro Area!!Estimate","Aguadilla-Isabela, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate","Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Median household income in the past 12 months ...,"44,799","±1,146","15,616",±344,"51,160",±559,"37,404","±1,010","45,644",...,"44,749","±1,242","58,269",±741,"42,609",±444,"49,781","±1,809","40,743","±1,111"


In [34]:
median_income2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Columns: 777 entries, Label (Grouping) to Yuma, AZ Metro Area!!Margin of Error
dtypes: object(777)
memory usage: 6.2+ KB


In [35]:
median_income2015.columns

Index(['Label (Grouping)', 'Abilene, TX Metro Area!!Estimate',
       'Abilene, TX Metro Area!!Margin of Error',
       'Aguadilla-Isabela, PR Metro Area!!Estimate',
       'Aguadilla-Isabela, PR Metro Area!!Margin of Error',
       'Akron, OH Metro Area!!Estimate',
       'Akron, OH Metro Area!!Margin of Error',
       'Albany, GA Metro Area!!Estimate',
       'Albany, GA Metro Area!!Margin of Error',
       'Albany, OR Metro Area!!Estimate',
       ...
       'Yakima, WA Metro Area!!Estimate',
       'Yakima, WA Metro Area!!Margin of Error',
       'York-Hanover, PA Metro Area!!Estimate',
       'York-Hanover, PA Metro Area!!Margin of Error',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error',
       'Yuba City, CA Metro Area!!Estimate',
       'Yuba City, CA Metro Area!!Margin of Error',
       'Yuma, AZ Metro Area!!Estimate',
       'Yuma, AZ Metro Area!!Margin of Error'],
      dtype='object', len

Population datasets B01003 from Census

#### Importing Population 2010 Dataset

In [36]:
Pop_2010 = pd.read_csv('Data Raw/Pop_2010.csv')

In [37]:
Pop_2010.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Estimate","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany-Schenectady-Troy, NY Metro Area!!Estimate",...,"Yauco, PR Metro Area!!Estimate","Yauco, PR Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate","Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Total,"163,092",*****,"309,197",*****,"703,093",*****,"157,528",*****,"865,982",...,"109,937",*****,"428,175",*****,"571,975",*****,"164,580",*****,"190,526",*****


In [38]:
Pop_2010.columns

Index(['Label (Grouping)', 'Abilene, TX Metro Area!!Estimate',
       'Abilene, TX Metro Area!!Margin of Error',
       'Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Estimate',
       'Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Margin of Error',
       'Akron, OH Metro Area!!Estimate',
       'Akron, OH Metro Area!!Margin of Error',
       'Albany, GA Metro Area!!Estimate',
       'Albany, GA Metro Area!!Margin of Error',
       'Albany-Schenectady-Troy, NY Metro Area!!Estimate',
       ...
       'Yauco, PR Metro Area!!Estimate',
       'Yauco, PR Metro Area!!Margin of Error',
       'York-Hanover, PA Metro Area!!Estimate',
       'York-Hanover, PA Metro Area!!Margin of Error',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate',
       'Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error',
       'Yuba City, CA Metro Area!!Estimate',
       'Yuba City, CA Metro Area!!Margin of Error',
       'Yuma, AZ Metro Area!!Estimate',
       'Yuma, AZ Metro Area!!M

In [39]:
Pop_2010.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Columns: 749 entries, Label (Grouping) to Yuma, AZ Metro Area!!Margin of Error
dtypes: object(749)
memory usage: 6.0+ KB


B15003 - Census datasets for Education Attainment over 25

#### Importing Education 2010 Dataset

In [40]:
Edu_2010 = pd.read_csv('Data Raw/Education_2010.csv')

In [41]:
Edu_2010.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Estimate","Aguadilla-Isabela-San Sebasti?n, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany-Schenectady-Troy, NY Metro Area!!Estimate",...,"Yauco, PR Metro Area!!Estimate","Yauco, PR Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren-Boardman, OH-PA Metro Area!!Estimate","Youngstown-Warren-Boardman, OH-PA Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Total:,"104,387","±1,049","203,050","±1,036","471,659","±1,401","103,864","±2,668","588,842",...,"69,273","±1,151","297,234",±746,"393,821",±973,"103,836","±1,301","120,795",±342
1,No schooling completed,"1,383",±763,"5,164","±1,018","3,145",±667,"1,832",±742,"4,415",...,"1,688",±706,"2,537",±774,"2,167",±588,"2,254",±553,"2,357",±727
2,Nursery school,15,±27,0,±285,0,±269,116,±138,0,...,68,±113,0,±246,38,±62,46,±78,88,±151
3,Kindergarten,0,±294,398,±262,136,±226,0,±300,124,...,0,±285,70,±114,39,±66,0,±294,0,±305
4,1st grade,122,±151,"1,471",±640,0,±269,80,±132,151,...,451,±328,0,±246,73,±94,0,±294,911,±632


In [42]:
Edu_2010.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Columns: 749 entries, Label (Grouping) to Yuma, AZ Metro Area!!Margin of Error
dtypes: object(749)
memory usage: 146.4+ KB


In [43]:
Edu_2010.shape

(25, 749)

In [44]:
import os

In [45]:
os.getcwd()

'/Users/timothyvermette/Desktop/Class - Jupyter Notebooks : ML:AI/Capstone Project'

#### Saving zillow_long to csv file. Dataset has been melted

In [46]:
zillow.shape

(895, 321)

In [47]:
zillow_long.shape

(282820, 8)

In [48]:
zillow_long.to_csv(
    'zillow_long.csv',
    index=False
)

In [49]:
zillow_long

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
0,102001,0,United States,country,NaN,2000-01-31,123427.049538,2000
1,394913,1,"New York, NY",msa,NY,2000-01-31,217908.609033,2000
2,753899,2,"Los Angeles, CA",msa,CA,2000-01-31,225552.897453,2000
3,394463,3,"Chicago, IL",msa,IL,2000-01-31,155258.141365,2000
4,394514,4,"Dallas, TX",msa,TX,2000-01-31,130734.453703,2000
...,...,...,...,...,...,...,...,...
282815,753929,935,"Zapata, TX",msa,TX,2026-04-30,125226.746254,2026
282816,394743,936,"Ketchikan, AK",msa,AK,2026-04-30,399056.055088,2026
282817,753874,937,"Craig, CO",msa,CO,2026-04-30,305274.568594,2026
282818,395188,938,"Vernon, TX",msa,TX,2026-04-30,103952.124110,2026


In [50]:
zillow_msa.to_csv(
    'zillow_msa.csv',
    index=False
)

#### Cleaning Median Income 2024 Dataset and naming copy as median_income2024_tidy.
##### Transposing, resetting Index, Changing column names to 'Metro' and 'Income2024'.
##### Replacing "Metro Area!!" and "Estimate" from 'Metro' to prepare to merge.

In [51]:
median_income2024_tidy = median_income2024.copy()

In [52]:
median_income2024_tidy.shape

(1, 787)

In [53]:
median_income2024_tidy = median_income2024_tidy.T

In [54]:
median_income2024_tidy.head()

,0
Label (Grouping),Median household income in the past 12 months ...
"Abilene, TX Metro Area!!Estimate","63,390"
"Abilene, TX Metro Area!!Margin of Error","±7,471"
"Aguadilla, PR Metro Area!!Estimate","24,351"
"Aguadilla, PR Metro Area!!Margin of Error","±1,426"


In [55]:
median_income2024_tidy.shape

(787, 1)

In [56]:
median_income2024_tidy = median_income2024_tidy.reset_index()

In [57]:
median_income2024_tidy.head()

,index,0
0,Label (Grouping),Median household income in the past 12 months ...
1,"Abilene, TX Metro Area!!Estimate","63,390"
2,"Abilene, TX Metro Area!!Margin of Error","±7,471"
3,"Aguadilla, PR Metro Area!!Estimate","24,351"
4,"Aguadilla, PR Metro Area!!Margin of Error","±1,426"


In [58]:
median_income2024_tidy.columns

Index(['index', 0], dtype='object')

In [59]:
median_income2024_tidy.columns = ['Metro', 'Income2024']

In [60]:
median_income2024_tidy.head()

,Metro,Income2024
0,Label (Grouping),Median household income in the past 12 months ...
1,"Abilene, TX Metro Area!!Estimate","63,390"
2,"Abilene, TX Metro Area!!Margin of Error","±7,471"
3,"Aguadilla, PR Metro Area!!Estimate","24,351"
4,"Aguadilla, PR Metro Area!!Margin of Error","±1,426"


In [61]:
median_income2024_tidy = median_income2024_tidy.iloc[1:].copy()

In [62]:
median_income2024_tidy = median_income2024_tidy[
    median_income2024_tidy['Metro'].str.contains('Estimate')
].copy()

In [63]:
median_income2024_tidy['Metro'] = (
    median_income2024_tidy['Metro']
    .str.replace(' Metro Area!!Estimate', '', regex=False)
)

In [64]:
median_income2024_tidy['Income2024'] = (
    median_income2024_tidy['Income2024']
    .str.replace(',', '', regex=False)
    .astype(float)
)

In [65]:
median_income2024_tidy.head()

,Metro,Income2024
1,"Abilene, TX",63390.0
3,"Aguadilla, PR",24351.0
5,"Akron, OH",71364.0
7,"Albany, GA",56851.0
9,"Albany, OR",80422.0


In [66]:
median_income2024_tidy.shape

(393, 2)

In [67]:
median_income2024_tidy = median_income2024_tidy.reset_index(drop=True)

#### Saving median_income2024_tidy.csv

In [68]:
median_income2024_tidy.to_csv('Data Intermediate/median_income2024_tidy.csv', index=False)

#### Creating zillow_2024
##### Filtering by Year 2024
##### Renaming 'RegionName' column to 'Metro'

In [69]:
zillow_2024 = zillow_msa[zillow_msa['Year'] == 2024].copy()
latest_2024_date = zillow_2024['Date'].max()
zillow_2024 = zillow_2024[zillow_2024['Date'] == latest_2024_date].copy()
zillow_2024 = zillow_2024[['RegionName', 'StateName', 'SizeRank', 'HomeValue']].copy()
zillow_2024 = zillow_2024.rename(columns={'RegionName': 'Metro'})

In [70]:
zillow_2024.head()

,Metro,StateName,SizeRank,HomeValue
267606,"New York, NY",NY,1,683398.809012
267607,"Los Angeles, CA",CA,2,980928.090661
267608,"Chicago, IL",IL,3,331370.219996
267609,"Dallas, TX",TX,4,385136.186667
267610,"Houston, TX",TX,5,316340.099352


In [71]:
zillow_2024.shape

(894, 4)

#### Merging zillow_2024 with median_income2024_tidy

In [72]:
merged_2024 = zillow_2024.merge(
    median_income2024_tidy,
    on='Metro',
    how='inner')

In [73]:
merged_2024.head()

,Metro,StateName,SizeRank,HomeValue,Income2024
0,"Pittsburgh, PA",PA,27,224138.021808,77214.0
1,"Columbus, OH",OH,32,324134.748480,82938.0
2,"Cleveland, OH",OH,35,234105.565617,72532.0
3,"Jacksonville, FL",FL,40,361318.502382,82053.0
4,"Oklahoma City, OK",OK,42,242392.335244,72930.0


In [74]:
merged_2024.shape

(215, 5)

#### Creating Feature of 'PriceToIncomeRatio' on merged dataset merged_2024

In [75]:
merged_2024['PriceToIncomeRatio'] = (merged_2024['HomeValue']/merged_2024['Income2024'])

In [76]:
merged_2024

,Metro,StateName,SizeRank,HomeValue,Income2024,PriceToIncomeRatio
0,"Pittsburgh, PA",PA,27,224138.021808,77214.0,2.902816
1,"Columbus, OH",OH,32,324134.748480,82938.0,3.908157
2,"Cleveland, OH",OH,35,234105.565617,72532.0,3.227618
3,"Jacksonville, FL",FL,40,361318.502382,82053.0,4.403477
4,"Oklahoma City, OK",OK,42,242392.335244,72930.0,3.323630
...,...,...,...,...,...,...
210,"Sandusky, OH",OH,469,226698.742262,74627.0,3.037758
211,"Enid, OK",OK,523,134741.055743,67004.0,2.010940
212,"Walla Walla, WA",WA,525,409039.620189,72892.0,5.611585
213,"Carson City, NV",NV,551,485634.763268,62948.0,7.714856


In [77]:
merged_2024[['Metro', 'HomeValue', 'Income2024', 'PriceToIncomeRatio']].head()

,Metro,HomeValue,Income2024,PriceToIncomeRatio
0,"Pittsburgh, PA",224138.021808,77214.0,2.902816
1,"Columbus, OH",324134.748480,82938.0,3.908157
2,"Cleveland, OH",234105.565617,72532.0,3.227618
3,"Jacksonville, FL",361318.502382,82053.0,4.403477
4,"Oklahoma City, OK",242392.335244,72930.0,3.323630


#### Checking and verifying quality of merge. Since merge took place on 'Metro' I am exploring how effective this was. 

In [78]:
zillow_metros = set(zillow_2024['Metro'])
income_metros = set(median_income2024_tidy['Metro'])
len(zillow_metros - income_metros)

679

In [79]:
len(income_metros - zillow_metros)

178

In [80]:
list(zillow_metros - income_metros)[:20]

['Dyersburg, TN',
 'Corning, NY',
 'Staunton, VA',
 'Mason City, IA',
 'Stevens Point, WI',
 'Dixon, IL',
 'Wenatchee, WA',
 'Marinette, WI',
 'Philadelphia, PA',
 'Enterprise, AL',
 'Yankton, SD',
 'Austin, MN',
 'Alpena, MI',
 'Whitewater, WI',
 'Sanford, NC',
 'Angola, IN',
 'Martinsville, VA',
 'Crescent City, CA',
 'Cordele, GA',
 'Grand Rapids, MI']

In [81]:
list(income_metros - zillow_metros)[:20]

['Duluth, MN-WI',
 'Charleston-North Charleston, SC',
 'Stockton-Lodi, CA',
 'Burlington-South Burlington, VT',
 'Durham-Chapel Hill, NC',
 'Manchester-Nashua, NH',
 'Youngstown-Warren, OH',
 'Kansas City, MO-KS',
 'Memphis, TN-MS-AR',
 'Arecibo, PR',
 'Florence-Muscle Shoals, AL',
 'San Jose-Sunnyvale-Santa Clara, CA',
 'Albany-Schenectady-Troy, NY',
 'Palm Bay-Melbourne-Titusville, FL',
 'Davenport-Moline-Rock Island, IA-IL',
 'Lansing-East Lansing, MI',
 'Columbus, GA-AL',
 'Lewiston-Auburn, ME',
 'Portland-South Portland, ME',
 'Chicago-Naperville-Elgin, IL-IN']

#### Saving merged_2024.csv to files

In [82]:
merged_2024.to_csv('Data Intermediate/merged_2024.csv')

#### Deeper dive into merge quality, investigating differences in Metros from top100_zillow, compared to top100_merged.

In [83]:
top100_zillow = zillow_2024[zillow_2024['SizeRank'] <= 100].copy()
top100_merged = top100_zillow.merge(
    median_income2024_tidy,
    on='Metro',
    how='inner'
)
top100_zillow.shape, top100_merged.shape

((97, 4), (29, 5))

#### It is evident that many major metropolitan cities were not merged.

In [84]:
top100_missing = set(top100_zillow['Metro']) - set(top100_merged['Metro'])
sorted(top100_missing)

['Albany, NY',
 'Allentown, PA',
 'Atlanta, GA',
 'Augusta, GA',
 'Austin, TX',
 'Bakersfield, CA',
 'Baltimore, MD',
 'Boston, MA',
 'Bridgeport, CT',
 'Buffalo, NY',
 'Cape Coral, FL',
 'Charleston, SC',
 'Charlotte, NC',
 'Chicago, IL',
 'Cincinnati, OH',
 'Dallas, TX',
 'Deltona, FL',
 'Denver, CO',
 'Des Moines, IA',
 'Detroit, MI',
 'Durham, NC',
 'Grand Rapids, MI',
 'Greensboro, NC',
 'Greenville, SC',
 'Harrisburg, PA',
 'Hartford, CT',
 'Houston, TX',
 'Indianapolis, IN',
 'Kansas City, MO',
 'Lakeland, FL',
 'Las Vegas, NV',
 'Little Rock, AR',
 'Los Angeles, CA',
 'Louisville, KY',
 'McAllen, TX',
 'Memphis, TN',
 'Miami, FL',
 'Milwaukee, WI',
 'Minneapolis, MN',
 'Nashville, TN',
 'New Orleans, LA',
 'New York, NY',
 'North Port, FL',
 'Omaha, NE',
 'Orlando, FL',
 'Oxnard, CA',
 'Palm Bay, FL',
 'Philadelphia, PA',
 'Phoenix, AZ',
 'Portland, OR',
 'Providence, RI',
 'Provo, UT',
 'Raleigh, NC',
 'Riverside, CA',
 'Sacramento, CA',
 'Salt Lake City, UT',
 'San Antonio, T

In [85]:
top100_zillow

,Metro,StateName,SizeRank,HomeValue
267606,"New York, NY",NY,1,683398.809012
267607,"Los Angeles, CA",CA,2,980928.090661
267608,"Chicago, IL",IL,3,331370.219996
267609,"Dallas, TX",TX,4,385136.186667
267610,"Houston, TX",TX,5,316340.099352
...,...,...,...,...
267698,"Augusta, GA",GA,96,241644.303151
267699,"Palm Bay, FL",FL,97,356479.087073
267700,"Jackson, MS",MS,98,203945.784396
267701,"Harrisburg, PA",PA,99,295320.038930


In [86]:
top100_merged

,Metro,StateName,SizeRank,HomeValue,Income2024
0,"Pittsburgh, PA",PA,27,224138.021808,77214.0
1,"Columbus, OH",OH,32,324134.748480,82938.0
2,"Cleveland, OH",OH,35,234105.565617,72532.0
3,"Jacksonville, FL",FL,40,361318.502382,82053.0
4,"Oklahoma City, OK",OK,42,242392.335244,72930.0
5,"Richmond, VA",VA,45,384723.855775,83460.0
6,"Birmingham, AL",AL,51,255817.918171,74954.0
7,"Rochester, NY",NY,52,265658.931546,76453.0
8,"Tucson, AZ",AZ,54,353589.494051,72067.0
9,"Urban Honolulu, HI",HI,55,855172.106000,105205.0


#### It becomes evident that Zillow and Census datasets use a differnet naming methodology for Metros.

In [87]:
top100_zillow['Metro'].sort_values().head(30)

267686               Akron, OH
267668              Albany, NY
267666         Albuquerque, NM
267674           Allentown, PA
267614             Atlanta, GA
267698             Augusta, GA
267634              Austin, TX
267667         Bakersfield, CA
267625           Baltimore, MD
267670         Baton Rouge, LA
267655          Birmingham, AL
267682          Boise City, ID
267615              Boston, MA
267664          Bridgeport, CT
267654             Buffalo, NY
267681          Cape Coral, FL
267678          Charleston, SC
267628           Charlotte, NC
267608             Chicago, IL
267633          Cincinnati, OH
267639           Cleveland, OH
267683    Colorado Springs, CO
267676            Columbia, SC
267637            Columbus, OH
267609              Dallas, TX
267692             Deltona, FL
267624              Denver, CO
267687          Des Moines, IA
267619             Detroit, MI
267697              Durham, NC
Name: Metro, dtype: object

In [89]:
median_income2024_tidy['Metro'].sort_values().head(30)

0                           Abilene, TX
1                         Aguadilla, PR
2                             Akron, OH
3                            Albany, GA
4                            Albany, OR
5           Albany-Schenectady-Troy, NY
6                       Albuquerque, NM
7                        Alexandria, LA
8     Allentown-Bethlehem-Easton, PA-NJ
9                           Altoona, PA
10                         Amarillo, TX
11                             Ames, IA
12         Amherst Town-Northampton, MA
13                        Anchorage, AK
14                        Ann Arbor, MI
15                  Anniston-Oxford, AL
16                         Appleton, WI
17                          Arecibo, PR
18                        Asheville, NC
19             Athens-Clarke County, GA
20    Atlanta-Sandy Springs-Roswell, GA
21          Atlantic City-Hammonton, NJ
22                   Auburn-Opelika, AL
23       Augusta-Richmond County, GA-SC
24     Austin-Round Rock-San Marcos, TX


#### Creating dataframe for mapping metros that will be used to complete this merge, and for our subsequent 2010 and 2015 datasets.
#### This metro_mapping dataframe will consist of columns 'ZillowMetro' and 'CensusMetro' and manual work will be completed to build dataframe for top 100 metros.

#### This will prove to be the lenghiest, most challenging, and time demanding part of preparing these datasets for merging. 

#### Below you will find several iterations of metro mapping, attempts to merge datasets, realizing there are more Metros missing, and more leg-work being completed until satisfaction with top 100 metros merged.

In [90]:
metro_mapping = pd.DataFrame(columns=[
    'ZillowMetro',
    'CensusMetro'
])

metro_mapping

,ZillowMetro,CensusMetro


In [91]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Dallas', case=False, na=False)
]

,Metro,Income2024
86,"Dallas-Fort Worth-Arlington, TX",92733.0


In [92]:
metro_mapping = pd.DataFrame({
    'ZillowMetro': ['Dallas, TX'],
    'CensusMetro': ['Dallas-Fort Worth-Arlington, TX']
})

metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"


In [93]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Los', case=False, na=False)
]

,Metro,Income2024
214,"Los Angeles-Long Beach-Anaheim, CA",96405.0


In [94]:
metro_mapping.loc[len(metro_mapping)] = [
    'Los Angeles, CA',
    'Los Angeles-Long Beach-Anaheim, CA'
]

In [95]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"


In [96]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('New York', case=False, na=False)
]

,Metro,Income2024
253,"New York-Newark-Jersey City, NY-NJ",99852.0


In [97]:
metro_mapping.loc[len(metro_mapping)] = [
    'New York, NY',
    'New York-Newark-Jersey City, NY-NJ'
]

In [98]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"


In [99]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Chicago', case=False, na=False)
]

,Metro,Income2024
69,"Chicago-Naperville-Elgin, IL-IN",90770.0


In [100]:
metro_mapping.loc[len(metro_mapping)] = [
    'Chicago, IL',
    'Chicago-Naperville-Elgin, IL-IN'
]

In [101]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"


In [102]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Atlanta', case=False, na=False)
]

,Metro,Income2024
20,"Atlanta-Sandy Springs-Roswell, GA",92344.0


In [103]:
metro_mapping.loc[len(metro_mapping)] = [
    'Atlanta, GA',
    'Atlanta-Sandy Springs-Roswell, GA'
]

In [104]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"


In [105]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Houston', case=False, na=False)
]

,Metro,Income2024
160,"Houston-Pasadena-The Woodlands, TX",81417.0


In [106]:
metro_mapping.loc[len(metro_mapping)] = [
    'Houston, TX',
    'Houston-Pasadena-The Woodlands, TX'
]

In [107]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"


In [108]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Dallas', case=False, na=False)
]

,Metro,Income2024
86,"Dallas-Fort Worth-Arlington, TX",92733.0


In [109]:
metro_mapping.loc[len(metro_mapping)] = [
    'Dallas, TX',
    'Dallas-Fort Worth-Arlington, TX'
]

In [110]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
6,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"


In [111]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Las Vegas', case=False, na=False)
]

,Metro,Income2024
200,"Las Vegas-Henderson-North Las Vegas, NV",80028.0


In [112]:
metro_mapping.loc[len(metro_mapping)] = [
    'Las Vegas, NV',
    'Las Vegas-Henderson-North Las Vegas, NV'
]

In [113]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Baltimore', case=False, na=False)
]

,Metro,Income2024
26,"Baltimore-Columbia-Towson, MD",98666.0


In [114]:
metro_mapping.loc[len(metro_mapping)] = [
    'Baltimore, MD',
    'Baltimore-Columbia-Towson, MD'
]

In [115]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('San Diego', case=False, na=False)
]

,Metro,Income2024
314,"San Diego-Chula Vista-Carlsbad, CA",109132.0


In [116]:
metro_mapping.loc[len(metro_mapping)] = [
    'San Diego, CA',
    'San Diego-Chula Vista-Carlsbad, CA'
]

In [117]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Denver', case=False, na=False)
]

,Metro,Income2024
94,"Denver-Aurora-Centennial, CO",108046.0


In [118]:
metro_mapping.loc[len(metro_mapping)] = [
    'Denver, CO',
    'Denver-Aurora-Centennial, CO'
]

In [119]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Miami', case=False, na=False)
]

,Metro,Income2024
229,"Miami-Fort Lauderdale-West Palm Beach, FL",80625.0


In [120]:
metro_mapping.loc[len(metro_mapping)] = [
    'Miami, FL',
    'Miami-Fort Lauderdale-West Palm Beach, FL'
]

In [121]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
6,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
7,"Las Vegas, NV","Las Vegas-Henderson-North Las Vegas, NV"
8,"Baltimore, MD","Baltimore-Columbia-Towson, MD"
9,"San Diego, CA","San Diego-Chula Vista-Carlsbad, CA"


In [122]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Boston', case=False, na=False)
]

,Metro,Income2024
44,"Boston-Cambridge-Newton, MA-NH",117825.0


In [123]:
metro_mapping.loc[len(metro_mapping)] = [
    'Boston, MA',
    'Boston-Cambridge-Newton, MA-NH'
]

In [124]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Seattle', case=False, na=False)
]

,Metro,Income2024
326,"Seattle-Tacoma-Bellevue, WA",112388.0


In [125]:
metro_mapping.loc[len(metro_mapping)] = [
    'Seattle, WA',
    'Seattle-Tacoma-Bellevue, WA'
]

In [126]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Philadelphia', case=False, na=False)
]

,Metro,Income2024
273,"Philadelphia-Camden-Wilmington, PA-NJ-DE-MD",90850.0


In [127]:
metro_mapping.loc[len(metro_mapping)] = [
    'Philadelphia, PA',
    'Philadelphia-Camden-Wilmington, PA-NJ-DE-MD'
]

In [128]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
6,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
7,"Las Vegas, NV","Las Vegas-Henderson-North Las Vegas, NV"
8,"Baltimore, MD","Baltimore-Columbia-Towson, MD"
9,"San Diego, CA","San Diego-Chula Vista-Carlsbad, CA"


In [131]:
metro_mapping = metro_mapping.drop_duplicates()

In [132]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
7,"Las Vegas, NV","Las Vegas-Henderson-North Las Vegas, NV"
8,"Baltimore, MD","Baltimore-Columbia-Towson, MD"
9,"San Diego, CA","San Diego-Chula Vista-Carlsbad, CA"
10,"Denver, CO","Denver-Aurora-Centennial, CO"


In [133]:
metro_mapping.to_csv(
    "Data Intermediate/metro_mapping.csv",
    index=False
)

In [134]:
mapping_dict = (
    metro_mapping
    .set_index('ZillowMetro')['CensusMetro']
    .to_dict()
)

In [135]:
top100_zillow_mapped = top100_zillow.copy()

In [136]:
top100_zillow_mapped['CensusMetro'] = (
    top100_zillow_mapped['Metro']
    .replace(mapping_dict)
)

In [137]:
top100_zillow_mapped[
    ['Metro', 'CensusMetro']
].head(20)

,Metro,CensusMetro
267606,"New York, NY","New York-Newark-Jersey City, NY-NJ"
267607,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
267608,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
267609,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
267610,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
267611,"Washington, DC","Washington, DC"
267612,"Philadelphia, PA","Philadelphia-Camden-Wilmington, PA-NJ-DE-MD"
267613,"Miami, FL","Miami-Fort Lauderdale-West Palm Beach, FL"
267614,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
267615,"Boston, MA","Boston-Cambridge-Newton, MA-NH"


In [138]:
income2024_for_merge = (
    median_income2024_tidy
    .rename(columns={'Metro': 'CensusMetro'})
    .copy()
)

In [139]:
top100_merged_mapped = top100_zillow_mapped.merge(
    income2024_for_merge,
    on='CensusMetro',
    how='left',
    validate='one_to_one'
)

In [140]:
top100_merged_mapped['Income2024'].notna().sum()

np.int64(43)

In [141]:
top100_merged_mapped.shape

(97, 6)

In [142]:
remaining_unmatched = (
    top100_merged_mapped[
        top100_merged_mapped['Income2024'].isna()
    ]
    [['Metro', 'CensusMetro', 'SizeRank']]
    .sort_values('SizeRank')
    .reset_index(drop=True)
)

remaining_unmatched

,Metro,CensusMetro,SizeRank
0,"Washington, DC","Washington, DC",6
1,"Phoenix, AZ","Phoenix, AZ",11
2,"San Francisco, CA","San Francisco, CA",12
3,"Riverside, CA","Riverside, CA",13
4,"Detroit, MI","Detroit, MI",14
5,"Minneapolis, MN","Minneapolis, MN",16
6,"Tampa, FL","Tampa, FL",18
7,"St. Louis, MO","St. Louis, MO",21
8,"Orlando, FL","Orlando, FL",22
9,"Charlotte, NC","Charlotte, NC",23


In [143]:
median_income2024_tidy[
    median_income2024_tidy['Metro']
    .str.contains('Washington', case=False, na=False)
]

,Metro,Income2024
372,"Washington-Arlington-Alexandria, DC-VA-MD-WV",126244.0


In [144]:
metro_mapping.loc[len(metro_mapping)] = [
    'Washington, DC',
    'Washington-Arlington-Alexandria, DC-VA-MD-WV'
]

/var/folders/4j/77_kc_5s4dgdv1jjvg0m69f40000gn/T/ipykernel_21314/1785436359.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  metro_mapping.loc[len(metro_mapping)] = [


In [145]:
metro_mapping = metro_mapping.copy()

In [146]:
metro_mapping.loc[len(metro_mapping)] = [
    'Washington, DC',
    'Washington-Arlington-Alexandria, DC-VA-MD-WV'
]

In [147]:
metro_mapping.tail()

,ZillowMetro,CensusMetro
10,"Denver, CO","Denver-Aurora-Centennial, CO"
11,"Miami, FL","Miami-Fort Lauderdale-West Palm Beach, FL"
12,"Boston, MA","Boston-Cambridge-Newton, MA-NH"
13,"Seattle, WA","Seattle-Tacoma-Bellevue, WA"
14,"Washington, DC","Washington-Arlington-Alexandria, DC-VA-MD-WV"


In [148]:
cities = ['Phoenix', 'San Francisco', 'Riverside', 'Detroit', 'Minneapolis']

for city in cities:
    print(f'\n{city}:')
    display(
        median_income2024_tidy[
            median_income2024_tidy['Metro']
            .str.contains(city, case=False, na=False)
        ]
    )


Phoenix:


,Metro,Income2024
274,"Phoenix-Mesa-Chandler, AZ",90133.0



San Francisco:


,Metro,Income2024
316,"San Francisco-Oakland-Fremont, CA",135590.0



Riverside:


,Metro,Income2024
295,"Riverside-San Bernardino-Ontario, CA",91013.0



Detroit:


,Metro,Income2024
96,"Detroit-Warren-Dearborn, MI",76403.0



Minneapolis:


,Metro,Income2024
234,"Minneapolis-St. Paul-Bloomington, MN-WI",97928.0


In [149]:
new_mappings = [
    ['Phoenix, AZ', 'Phoenix-Mesa-Chandler, AZ'],
    ['San Francisco, CA', 'San Francisco-Oakland-Fremont, CA'],
    ['Riverside, CA', 'Riverside-San Bernardino-Ontario, CA'],
    ['Detroit, MI', 'Detroit-Warren-Dearborn, MI'],
    ['Minneapolis, MN', 'Minneapolis-St. Paul-Bloomington, MN-WI']
]

In [150]:
new_mapping_df = pd.DataFrame(
    new_mappings,
    columns=['ZillowMetro', 'CensusMetro']
)

metro_mapping = pd.concat(
    [metro_mapping.copy(), new_mapping_df],
    ignore_index=True
)

metro_mapping = (
    metro_mapping
    .drop_duplicates(subset='ZillowMetro', keep='last')
    .reset_index(drop=True)
)

In [152]:
metro_mapping.to_csv(
    'Data Intermediate/metro_mapping.csv',
    index=False
)

In [153]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
6,"Las Vegas, NV","Las Vegas-Henderson-North Las Vegas, NV"
7,"Baltimore, MD","Baltimore-Columbia-Towson, MD"
8,"San Diego, CA","San Diego-Chula Vista-Carlsbad, CA"
9,"Denver, CO","Denver-Aurora-Centennial, CO"


In [156]:
cities = ['Orlando', 'Charlotte', 'Austin', 'San Antonio', 'Portland', 'Sacramento', 'Nashville', 'Kansas City', 'Cincinnati', 'Milwaukee', 'Indianapolis', 'St. Louis']

for city in cities:
    print(f'\n{city}:')
    display(
        median_income2024_tidy[
            median_income2024_tidy['Metro']
            .str.contains(city, case=False, na=False)
        ]
    )


Orlando:


,Metro,Income2024
263,"Orlando-Kissimmee-Sanford, FL",81044.0



Charlotte:


,Metro,Income2024
65,"Charlotte-Concord-Gastonia, NC-SC",85938.0
66,"Charlottesville, VA",95796.0



Austin:


,Metro,Income2024
24,"Austin-Round Rock-San Marcos, TX",99897.0



San Antonio:


,Metro,Income2024
313,"San Antonio-New Braunfels, TX",78112.0



Portland:


,Metro,Income2024
280,"Portland-South Portland, ME",93062.0
281,"Portland-Vancouver-Hillsboro, OR-WA",98994.0



Sacramento:


,Metro,Income2024
302,"Sacramento-Roseville-Folsom, CA",98775.0



Nashville:


,Metro,Income2024
250,"Nashville-Davidson--Murfreesboro--Franklin, TN",88800.0



Kansas City:


,Metro,Income2024
181,"Kansas City, MO-KS",83785.0



Cincinnati:


,Metro,Income2024
71,"Cincinnati, OH-KY-IN",81489.0



Milwaukee:


,Metro,Income2024
233,"Milwaukee-Waukesha, WI",77919.0



Indianapolis:


,Metro,Income2024
164,"Indianapolis-Carmel-Greenwood, IN",80239.0



St. Louis:


,Metro,Income2024
307,"St. Louis, MO-IL",81679.0


In [159]:
new_mappings = [
    ['Orlando, FL', 'Orlando-Kissimmee-Sanford, FL'],
    ['Charlotte, NC', 'Charlotte-Concord-Gastonia, NC-SC'],
    ['Austin, TX', 'Austin-Round Rock-San Marcos, TX'],
    ['San Antonio, TX', 'San Antonio-New Braunfels, TX'],
    ['Portland, OR', 'Portland-Vancouver-Hillsboro, OR-WA'],
    ['Sacramento, CA', 'Sacramento-Roseville-Folsom, CA'],
    ['Nashville, TN', 'Nashville-Davidson--Murfreesboro--Franklin, TN'],
    ['Kansas City, MO', 'Kansas City, MO-KS'],
    ['Cincinnati, OH', 'Cincinnati, OH-KY-IN'],    
    ['Milwaukee, WI', 'Milwaukee-Waukesha, WI'],
    ['Indianapolis, IN', 'Indianapolis-Carmel-Greenwood, IN'],
    ['St. Louis, MO', 'St. Louis, MO-IL']
]

In [160]:
new_mapping_df = pd.DataFrame(
    new_mappings,
    columns=['ZillowMetro', 'CensusMetro']
)

metro_mapping = pd.concat(
    [metro_mapping.copy(), new_mapping_df],
    ignore_index=True
)

metro_mapping = (
    metro_mapping
    .drop_duplicates(subset='ZillowMetro', keep='last')
    .reset_index(drop=True)
)

In [161]:
metro_mapping.to_csv(
    'Data Intermediate/metro_mapping.csv',
    index=False
)

In [162]:
print(f'Total mappings: {len(metro_mapping)}')
metro_mapping.tail()

Total mappings: 31


,ZillowMetro,CensusMetro
26,"Kansas City, MO","Kansas City, MO-KS"
27,"Cincinnati, OH","Cincinnati, OH-KY-IN"
28,"Milwaukee, WI","Milwaukee-Waukesha, WI"
29,"Indianapolis, IN","Indianapolis-Carmel-Greenwood, IN"
30,"St. Louis, MO","St. Louis, MO-IL"


In [163]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
5,"Houston, TX","Houston-Pasadena-The Woodlands, TX"
6,"Las Vegas, NV","Las Vegas-Henderson-North Las Vegas, NV"
7,"Baltimore, MD","Baltimore-Columbia-Towson, MD"
8,"San Diego, CA","San Diego-Chula Vista-Carlsbad, CA"
9,"Denver, CO","Denver-Aurora-Centennial, CO"


In [164]:
mapping_dict = (
    metro_mapping
    .set_index('ZillowMetro')['CensusMetro']
    .to_dict()
)

In [165]:
top100_zillow_mapped = top100_zillow.copy()

top100_zillow_mapped['CensusMetro'] = (
    top100_zillow_mapped['Metro']
    .replace(mapping_dict)
)

In [166]:
top100_merged_mapped = top100_zillow_mapped.merge(
    income2024_for_merge,
    on='CensusMetro',
    how='left',
    validate='one_to_one'
)

In [167]:
print(f"Matched metros: {top100_merged_mapped['Income2024'].notna().sum()}")

Matched metros: 60


In [168]:
remaining_unmatched = (
    top100_merged_mapped[
        top100_merged_mapped['Income2024'].isna()
    ][['Metro', 'CensusMetro', 'SizeRank']]
    .sort_values('SizeRank')
    .reset_index(drop=True)
)

remaining_unmatched

,Metro,CensusMetro,SizeRank
0,"Philadelphia, PA","Philadelphia, PA",7
1,"Tampa, FL","Tampa, FL",18
2,"San Jose, CA","San Jose, CA",36
3,"Virginia Beach, VA","Virginia Beach, VA",38
4,"Providence, RI","Providence, RI",39
5,"Raleigh, NC","Raleigh, NC",43
6,"Memphis, TN","Memphis, TN",44
7,"Louisville, KY","Louisville, KY",46
8,"New Orleans, LA","New Orleans, LA",47
9,"Salt Lake City, UT","Salt Lake City, UT",48


In [169]:
cities = [
    'Philadelphia',
    'Tampa',
    'San Jose',
    'Virginia Beach',
    'Providence'
]

for city in cities:
    print(f'\n{city}:')
    display(
        median_income2024_tidy[
            median_income2024_tidy['Metro']
            .str.contains(city, case=False, na=False)
        ]
    )


Philadelphia:


,Metro,Income2024
273,"Philadelphia-Camden-Wilmington, PA-NJ-DE-MD",90850.0



Tampa:


,Metro,Income2024
349,"Tampa-St. Petersburg-Clearwater, FL",78275.0



San Jose:


,Metro,Income2024
317,"San Jose-Sunnyvale-Santa Clara, CA",164801.0



Virginia Beach:


,Metro,Income2024
367,"Virginia Beach-Chesapeake-Norfolk, VA-NC",82402.0



Providence:


,Metro,Income2024
284,"Providence-Warwick, RI-MA",82870.0


In [170]:
metro_mapping[
    metro_mapping['ZillowMetro'] == 'Philadelphia, PA'
]

,ZillowMetro,CensusMetro


In [171]:
new_mappings = [
    ['Philadelphia, PA', 'Philadelphia-Camden-Wilmington, PA-NJ-DE-MD'],
    ['Tampa, FL', 'Tampa-St. Petersburg-Clearwater, FL'],
    ['San Jose, CA', 'San Jose-Sunnyvale-Santa Clara, CA'],
    ['Virginia Beach, VA', 'Virginia Beach-Chesapeake-Norfolk, VA-NC'],
    ['Providence, RI', 'Providence-Warwick, RI-MA']
]

In [172]:
new_mapping_df = pd.DataFrame(
    new_mappings,
    columns=['ZillowMetro', 'CensusMetro']
)

metro_mapping = pd.concat(
    [metro_mapping.copy(), new_mapping_df],
    ignore_index=True
)

metro_mapping = (
    metro_mapping
    .drop_duplicates(subset='ZillowMetro', keep='last')
    .reset_index(drop=True)
)

In [175]:
metro_mapping.to_csv(
    'Data Intermediate/metro_mapping.csv',
    index=False
)

In [176]:
mapping_dict = (
    metro_mapping
    .set_index('ZillowMetro')['CensusMetro']
    .to_dict()
)

In [177]:
top100_zillow_mapped = top100_zillow.copy()

top100_zillow_mapped['CensusMetro'] = (
    top100_zillow_mapped['Metro']
    .replace(mapping_dict)
)

In [178]:
top100_merged_mapped = top100_zillow_mapped.merge(
    income2024_for_merge,
    on='CensusMetro',
    how='left',
    validate='one_to_one'
)

In [179]:
print(top100_merged_mapped['Income2024'].notna().sum())

65


In [180]:
top100_merged_mapped.to_csv(
    'Data Intermediate/top100_merged_2024.csv',
    index=False
)

In [181]:
print(f"Total Zillow metros: {len(top100_merged_mapped)}")
print(f"Matched metros: {top100_merged_mapped['Income2024'].notna().sum()}")
print(f"Unmatched metros: {top100_merged_mapped['Income2024'].isna().sum()}")

Total Zillow metros: 97
Matched metros: 65
Unmatched metros: 32


In [182]:
zillow_msa.head()

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
1,394913,1,"New York, NY",msa,NY,2000-01-31,217908.609033,2000
2,753899,2,"Los Angeles, CA",msa,CA,2000-01-31,225552.897453,2000
3,394463,3,"Chicago, IL",msa,IL,2000-01-31,155258.141365,2000
4,394514,4,"Dallas, TX",msa,TX,2000-01-31,130734.453703,2000
5,394692,5,"Houston, TX",msa,TX,2000-01-31,125738.933364,2000


In [183]:
zillow_msa['Date'].unique()

<DatetimeArray>
['2000-01-31 00:00:00', '2000-02-29 00:00:00', '2000-03-31 00:00:00',
 '2000-04-30 00:00:00', '2000-05-31 00:00:00', '2000-06-30 00:00:00',
 '2000-07-31 00:00:00', '2000-08-31 00:00:00', '2000-09-30 00:00:00',
 '2000-10-31 00:00:00',
 ...
 '2025-07-31 00:00:00', '2025-08-31 00:00:00', '2025-09-30 00:00:00',
 '2025-10-31 00:00:00', '2025-11-30 00:00:00', '2025-12-31 00:00:00',
 '2026-01-31 00:00:00', '2026-02-28 00:00:00', '2026-03-31 00:00:00',
 '2026-04-30 00:00:00']
Length: 316, dtype: datetime64[ns]

In [184]:
zillow_msa['Date'].dt.year.unique()

array([2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
       2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021,
       2022, 2023, 2024, 2025, 2026], dtype=int32)

In [185]:
zillow_msa[
    zillow_msa['Year'] == 2024
].head()

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
257761,394913,1,"New York, NY",msa,NY,2024-01-31,641223.118800,2024
257762,753899,2,"Los Angeles, CA",msa,CA,2024-01-31,941095.228037,2024
257763,394463,3,"Chicago, IL",msa,IL,2024-01-31,315221.243164,2024
257764,394514,4,"Dallas, TX",msa,TX,2024-01-31,385775.892684,2024
257765,394692,5,"Houston, TX",msa,TX,2024-01-31,314847.568079,2024


In [186]:
zillow_msa[
    zillow_msa['Year'] == 2024
].shape

(10728, 8)

In [187]:
zillow_msa[
    zillow_msa['Year'] == 2024
]['Date'].unique()

<DatetimeArray>
['2024-01-31 00:00:00', '2024-02-29 00:00:00', '2024-03-31 00:00:00',
 '2024-04-30 00:00:00', '2024-05-31 00:00:00', '2024-06-30 00:00:00',
 '2024-07-31 00:00:00', '2024-08-31 00:00:00', '2024-09-30 00:00:00',
 '2024-10-31 00:00:00', '2024-11-30 00:00:00', '2024-12-31 00:00:00']
Length: 12, dtype: datetime64[ns]

In [188]:
zillow_2024 = zillow_msa[
    zillow_msa['Date'] == '2024-12-31'
].copy()

zillow_2024.shape

(894, 8)

In [189]:
zillow_2024.head()

,RegionID,SizeRank,RegionName,RegionType,StateName,Date,HomeValue,Year
267606,394913,1,"New York, NY",msa,NY,2024-12-31,683398.809012,2024
267607,753899,2,"Los Angeles, CA",msa,CA,2024-12-31,980928.090661,2024
267608,394463,3,"Chicago, IL",msa,IL,2024-12-31,331370.219996,2024
267609,394514,4,"Dallas, TX",msa,TX,2024-12-31,385136.186667,2024
267610,394692,5,"Houston, TX",msa,TX,2024-12-31,316340.099352,2024


In [190]:
zillow_2024 = zillow_2024.rename(
    columns={'RegionName': 'Metro'}
)

In [191]:
zillow_2024.to_csv(
    'Data Intermediate/zillow_2024.csv',
    index=False
)

#### Importing Population2024.csv, taking this dataset through the same steps as previous income dataset since census uses same metro naming methodology, transposing, resetting index, preparing for merge.

In [192]:
population2024 = pd.read_csv('Data Raw/Pop_2024.csv')

In [197]:
population2024.head()

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla, PR Metro Area!!Estimate","Aguadilla, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren, OH Metro Area!!Estimate","Youngstown-Warren, OH Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Total,"180,408",*****,"251,769",*****,"699,820",*****,"146,018",*****,"130,706",...,"257,152",*****,"462,924",*****,"427,420",*****,"183,364",*****,"211,741",*****


In [199]:
population2024.shape

(1, 787)

In [200]:
population2024.columns

Index(['Label (Grouping)', 'Abilene, TX Metro Area!!Estimate',
       'Abilene, TX Metro Area!!Margin of Error',
       'Aguadilla, PR Metro Area!!Estimate',
       'Aguadilla, PR Metro Area!!Margin of Error',
       'Akron, OH Metro Area!!Estimate',
       'Akron, OH Metro Area!!Margin of Error',
       'Albany, GA Metro Area!!Estimate',
       'Albany, GA Metro Area!!Margin of Error',
       'Albany, OR Metro Area!!Estimate',
       ...
       'Yakima, WA Metro Area!!Estimate',
       'Yakima, WA Metro Area!!Margin of Error',
       'York-Hanover, PA Metro Area!!Estimate',
       'York-Hanover, PA Metro Area!!Margin of Error',
       'Youngstown-Warren, OH Metro Area!!Estimate',
       'Youngstown-Warren, OH Metro Area!!Margin of Error',
       'Yuba City, CA Metro Area!!Estimate',
       'Yuba City, CA Metro Area!!Margin of Error',
       'Yuma, AZ Metro Area!!Estimate',
       'Yuma, AZ Metro Area!!Margin of Error'],
      dtype='object', length=787)

In [210]:
population2024_tidy = population2024.T.reset_index()

In [211]:
population2024_tidy.head(10)

,index,0
0,Label (Grouping),Total
1,"Abilene, TX Metro Area!!Estimate","180,408"
2,"Abilene, TX Metro Area!!Margin of Error",*****
3,"Aguadilla, PR Metro Area!!Estimate","251,769"
4,"Aguadilla, PR Metro Area!!Margin of Error",*****
5,"Akron, OH Metro Area!!Estimate","699,820"
6,"Akron, OH Metro Area!!Margin of Error",*****
7,"Albany, GA Metro Area!!Estimate","146,018"
8,"Albany, GA Metro Area!!Margin of Error",*****
9,"Albany, OR Metro Area!!Estimate","130,706"


In [215]:
population2024_tidy.columns

Index(['Metro', 'Population2024'], dtype='object')

In [216]:
population2024_tidy.columns = ['Metro', 'Population2024']

In [217]:
population2024_tidy.head()

,Metro,Population2024
0,Label (Grouping),Total
1,"Abilene, TX Metro Area!!Estimate","180,408"
2,"Abilene, TX Metro Area!!Margin of Error",*****
3,"Aguadilla, PR Metro Area!!Estimate","251,769"
4,"Aguadilla, PR Metro Area!!Margin of Error",*****


In [218]:
population2024_tidy = population2024_tidy.iloc[1:].copy()

In [219]:
population2024_tidy = population2024_tidy[
    population2024_tidy['Metro'].str.contains('Estimate')
].copy()

In [220]:
population2024_tidy['Metro'] = (
    population2024_tidy['Metro']
    .str.replace(' Metro Area!!Estimate', '', regex=False)
)

In [221]:
population2024_tidy['Population2024'] = (
    population2024_tidy['Population2024']
    .str.replace(',', '', regex=False)
    .astype(float)
)

In [222]:
population2024_tidy.head()

,Metro,Population2024
1,"Abilene, TX",180408.0
3,"Aguadilla, PR",251769.0
5,"Akron, OH",699820.0
7,"Albany, GA",146018.0
9,"Albany, OR",130706.0


In [224]:
population2024_tidy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 393 entries, 1 to 785
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Metro           393 non-null    object 
 1   Population2024  393 non-null    float64
dtypes: float64(1), object(1)
memory usage: 9.2+ KB


In [225]:
population2024_tidy.head()
population2024_tidy.shape
population2024_tidy['Population2024'].describe()

count    3.930000e+02
mean     7.429407e+05
std      1.649234e+06
min      5.808200e+04
25%      1.463860e+05
50%      2.469890e+05
75%      5.777260e+05
max      1.979854e+07
Name: Population2024, dtype: float64

In [226]:
population2024_tidy.to_csv(
    'Data Intermediate/population2024_tidy.csv',
    index=False
)

In [227]:
population2024_tidy = population2024_tidy.rename(
    columns={'Metro': 'CensusMetro'}
)

In [228]:
population2024_tidy.head()

,CensusMetro,Population2024
1,"Abilene, TX",180408.0
3,"Aguadilla, PR",251769.0
5,"Akron, OH",699820.0
7,"Albany, GA",146018.0
9,"Albany, OR",130706.0


In [229]:
zillow_2024

,RegionID,SizeRank,Metro,RegionType,StateName,Date,HomeValue,Year
267606,394913,1,"New York, NY",msa,NY,2024-12-31,683398.809012,2024
267607,753899,2,"Los Angeles, CA",msa,CA,2024-12-31,980928.090661,2024
267608,394463,3,"Chicago, IL",msa,IL,2024-12-31,331370.219996,2024
267609,394514,4,"Dallas, TX",msa,TX,2024-12-31,385136.186667,2024
267610,394692,5,"Houston, TX",msa,TX,2024-12-31,316340.099352,2024
...,...,...,...,...,...,...,...,...
268495,753929,935,"Zapata, TX",msa,TX,2024-12-31,135137.495449,2024
268496,394743,936,"Ketchikan, AK",msa,AK,2024-12-31,362746.087118,2024
268497,753874,937,"Craig, CO",msa,CO,2024-12-31,294932.601827,2024
268498,395188,938,"Vernon, TX",msa,TX,2024-12-31,110257.818754,2024


In [230]:
mapping_dict = (
    metro_mapping
    .set_index('ZillowMetro')['CensusMetro']
    .to_dict()
)

In [231]:
zillow_2024_mapped = zillow_2024.copy()

zillow_2024_mapped['CensusMetro'] = (
    zillow_2024_mapped['Metro']
    .replace(mapping_dict)
)

In [232]:
zillow_2024_mapped[
    ['Metro', 'CensusMetro', 'HomeValue']
].head(10)

,Metro,CensusMetro,HomeValue
267606,"New York, NY","New York-Newark-Jersey City, NY-NJ",683398.809012
267607,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA",980928.090661
267608,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN",331370.219996
267609,"Dallas, TX","Dallas-Fort Worth-Arlington, TX",385136.186667
267610,"Houston, TX","Houston-Pasadena-The Woodlands, TX",316340.099352
267611,"Washington, DC","Washington-Arlington-Alexandria, DC-VA-MD-WV",582933.450850
267612,"Philadelphia, PA","Philadelphia-Camden-Wilmington, PA-NJ-DE-MD",372959.178049
267613,"Miami, FL","Miami-Fort Lauderdale-West Palm Beach, FL",496754.730919
267614,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA",395244.663130
267615,"Boston, MA","Boston-Cambridge-Newton, MA-NH",722124.827928


In [233]:
income2024_for_merge = (
    median_income2024_tidy
    .rename(columns={'Metro': 'CensusMetro'})
    [['CensusMetro', 'Income2024']]
    .copy()
)

In [235]:
population2024_for_merge = population2024_tidy[
    ['CensusMetro', 'Population2024']
].copy()

In [236]:
analysis_2024 = zillow_2024_mapped.merge(
    income2024_for_merge,
    on='CensusMetro',
    how='left',
    validate='one_to_one'
)

In [237]:
analysis_2024 = analysis_2024.merge(
    population2024_for_merge,
    on='CensusMetro',
    how='left',
    validate='one_to_one'
)

In [238]:
analysis_2024_top100 = (
    analysis_2024
    .nsmallest(100, 'SizeRank')
    .copy()
)

In [239]:
analysis_2024_top100[
    [
        'SizeRank',
        'Metro',
        'CensusMetro',
        'HomeValue',
        'Income2024',
        'Population2024'
    ]
].head(20)

,SizeRank,Metro,CensusMetro,HomeValue,Income2024,Population2024
0,1,"New York, NY","New York-Newark-Jersey City, NY-NJ",6.833988e+05,99852.0,19798537.0
1,2,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA",9.809281e+05,96405.0,12974487.0
2,3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN",3.313702e+05,90770.0,9371595.0
3,4,"Dallas, TX","Dallas-Fort Worth-Arlington, TX",3.851362e+05,92733.0,7985590.0
4,5,"Houston, TX","Houston-Pasadena-The Woodlands, TX",3.163401e+05,81417.0,7442788.0
5,6,"Washington, DC","Washington-Arlington-Alexandria, DC-VA-MD-WV",5.829335e+05,126244.0,6318377.0
6,7,"Philadelphia, PA","Philadelphia-Camden-Wilmington, PA-NJ-DE-MD",3.729592e+05,90850.0,6272014.0
7,8,"Miami, FL","Miami-Fort Lauderdale-West Palm Beach, FL",4.967547e+05,80625.0,6249291.0
8,9,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA",3.952447e+05,92344.0,6256377.0
9,10,"Boston, MA","Boston-Cambridge-Newton, MA-NH",7.221248e+05,117825.0,4953354.0


In [240]:
print(
    'Income matches:',
    analysis_2024_top100['Income2024'].notna().sum()
)

print(
    'Population matches:',
    analysis_2024_top100['Population2024'].notna().sum()
)

print(
    'Complete rows:',
    analysis_2024_top100[
        ['Income2024', 'Population2024']
    ].notna().all(axis=1).sum()
)

Income matches: 66
Population matches: 66
Complete rows: 66


In [244]:
analysis_2024_top100.to_csv(
    'Data Intermediate/analysis_2024_top100_homevalue_income_population.csv',
    index=False
)

In [245]:
analysis_2024_top100[
    ['HomeValue', 'Income2024', 'Population2024']
].describe()

,HomeValue,Income2024,Population2024
count,1.000000e+02,66.000000,6.600000e+01
mean,4.163581e+05,87309.363636,2.937380e+06
std,2.213067e+05,17412.564528,3.189676e+06
min,1.911165e+05,59834.000000,4.628150e+05
25%,2.727046e+05,76335.500000,9.241718e+05
50%,3.613780e+05,82636.000000,2.150790e+06
75%,4.673956e+05,95487.000000,3.300712e+06
max,1.679506e+06,164801.000000,1.979854e+07


In [246]:
analysis_2024_top100[
    ['HomeValue', 'Income2024', 'Population2024']
].isna().sum()

HomeValue          0
Income2024        34
Population2024    34
dtype: int64

In [247]:
education2024 = pd.read_csv('Data Raw/Education_2024.csv')

In [248]:
education2024

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla, PR Metro Area!!Estimate","Aguadilla, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren, OH Metro Area!!Estimate","Youngstown-Warren, OH Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
0,Total:,"114,622",±382,"190,498",*****,"492,073",±435,"97,190",±309,"92,020",...,"157,998",±155,"325,365",±77,"306,607",±176,"118,976",±248,"137,936",±230
1,No schooling completed,"1,646",±331,"4,180",±716,"7,212",±751,"1,295",±290,687,...,"7,677",±694,"3,298",±449,"2,782",±390,"6,174",±725,"4,134",±598
2,Nursery school,36,±41,42,±52,55,±38,21,±26,14,...,36,±31,67,±75,94,±90,31,±35,17,±14
3,Kindergarten,0,±35,43,±68,68,±53,21,±25,12,...,171,±115,25,±37,39,±33,73,±67,7,±11
4,1st grade,68,±55,"1,052",±310,89,±85,40,±48,0,...,387,±264,88,±56,144,±125,0,±32,293,±174
5,2nd grade,151,±100,"1,253",±316,41,±48,129,±85,16,...,"1,214",±336,62,±61,38,±35,155,±139,593,±249
6,3rd grade,355,±200,"3,065",±411,102,±128,99,±101,29,...,819,±210,42,±28,271,±206,319,±134,"1,016",±303
7,4th grade,206,±109,"3,207",±478,120,±94,98,±79,57,...,728,±285,145,±88,128,±90,331,±157,717,±273
8,5th grade,228,±123,"3,412",±536,276,±138,94,±70,120,...,"1,136",±290,279,±132,298,±145,449,±184,885,±368
9,6th grade,841,±216,"6,366",±863,551,±193,425,±215,415,...,"5,547",±625,740,±249,467,±188,"1,398",±318,"4,367",±675


In [249]:
education2024_work = education2024.copy()

In [250]:
bachelors = education2024_work.iloc[[21]].copy()

In [251]:
bachelors

,Label (Grouping),"Abilene, TX Metro Area!!Estimate","Abilene, TX Metro Area!!Margin of Error","Aguadilla, PR Metro Area!!Estimate","Aguadilla, PR Metro Area!!Margin of Error","Akron, OH Metro Area!!Estimate","Akron, OH Metro Area!!Margin of Error","Albany, GA Metro Area!!Estimate","Albany, GA Metro Area!!Margin of Error","Albany, OR Metro Area!!Estimate",...,"Yakima, WA Metro Area!!Estimate","Yakima, WA Metro Area!!Margin of Error","York-Hanover, PA Metro Area!!Estimate","York-Hanover, PA Metro Area!!Margin of Error","Youngstown-Warren, OH Metro Area!!Estimate","Youngstown-Warren, OH Metro Area!!Margin of Error","Yuba City, CA Metro Area!!Estimate","Yuba City, CA Metro Area!!Margin of Error","Yuma, AZ Metro Area!!Estimate","Yuma, AZ Metro Area!!Margin of Error"
21,Bachelor's degree,"19,850","±1,119","34,595","±1,628","107,457","±2,143","11,855",±891,"13,606",...,"19,276","±1,129","56,117","±1,785","48,099","±1,514","15,720","±1,157","14,688","±1,160"


In [252]:
bachelors = bachelors.T.reset_index()

In [253]:
bachelors

,index,21
0,Label (Grouping),Bachelor's degree
1,"Abilene, TX Metro Area!!Estimate","19,850"
2,"Abilene, TX Metro Area!!Margin of Error","±1,119"
3,"Aguadilla, PR Metro Area!!Estimate","34,595"
4,"Aguadilla, PR Metro Area!!Margin of Error","±1,628"
...,...,...
782,"Youngstown-Warren, OH Metro Area!!Margin of Error","±1,514"
783,"Yuba City, CA Metro Area!!Estimate","15,720"
784,"Yuba City, CA Metro Area!!Margin of Error","±1,157"
785,"Yuma, AZ Metro Area!!Estimate","14,688"


In [254]:
bachelors.columns = ['Metro', 'Bachelors']

In [255]:
bachelors = bachelors.iloc[1:].copy()

In [256]:
bachelors = bachelors[
    bachelors['Metro'].str.contains('Estimate')
].copy()

In [257]:
bachelors

,Metro,Bachelors
1,"Abilene, TX Metro Area!!Estimate","19,850"
3,"Aguadilla, PR Metro Area!!Estimate","34,595"
5,"Akron, OH Metro Area!!Estimate","107,457"
7,"Albany, GA Metro Area!!Estimate","11,855"
9,"Albany, OR Metro Area!!Estimate","13,606"
...,...,...
777,"Yakima, WA Metro Area!!Estimate","19,276"
779,"York-Hanover, PA Metro Area!!Estimate","56,117"
781,"Youngstown-Warren, OH Metro Area!!Estimate","48,099"
783,"Yuba City, CA Metro Area!!Estimate","15,720"


In [260]:
bachelors['Metro'] = (
    bachelors['Metro']
    .str.replace(' Metro Area!!Estimate', '', regex=False)
)

In [261]:
bachelors

,Metro,Bachelors
1,"Abilene, TX","19,850"
3,"Aguadilla, PR","34,595"
5,"Akron, OH","107,457"
7,"Albany, GA","11,855"
9,"Albany, OR","13,606"
...,...,...
777,"Yakima, WA","19,276"
779,"York-Hanover, PA","56,117"
781,"Youngstown-Warren, OH","48,099"
783,"Yuba City, CA","15,720"


In [262]:
bachelors['Bachelors'] = (
    bachelors['Bachelors']
    .astype(str)
    .str.replace(',', '', regex=False)
)

bachelors['Bachelors'] = pd.to_numeric(bachelors['Bachelors'])

In [263]:
bachelors

,Metro,Bachelors
1,"Abilene, TX",19850
3,"Aguadilla, PR",34595
5,"Akron, OH",107457
7,"Albany, GA",11855
9,"Albany, OR",13606
...,...,...
777,"Yakima, WA",19276
779,"York-Hanover, PA",56117
781,"Youngstown-Warren, OH",48099
783,"Yuba City, CA",15720


In [264]:
def clean_education_row(df, row_number, column_name):
    """
    Extracts one education level from the ACS education table
    and returns a tidy Metro/value DataFrame.
    """

    temp = df.iloc[[row_number]].T.reset_index()

    temp.columns = ['Metro', column_name]

    temp = temp.iloc[1:].copy()

    temp = temp[
        temp['Metro'].str.contains('Estimate')
    ].copy()

    temp['Metro'] = (
        temp['Metro']
        .str.replace(' Metro Area!!Estimate', '', regex=False)
    )

    temp[column_name] = (
        temp[column_name]
        .astype(str)
        .str.replace(',', '', regex=False)
    )

    temp[column_name] = pd.to_numeric(temp[column_name])

    return temp

In [265]:
bachelors = clean_education_row(education2024, 21, 'Bachelors')

masters = clean_education_row(education2024, 22, 'Masters')

professional = clean_education_row(education2024, 23, 'Professional')

doctorate = clean_education_row(education2024, 24, 'Doctorate')

In [266]:
college2024 = bachelors.copy()

college2024 = college2024.merge(
    masters,
    on='Metro'
)

college2024 = college2024.merge(
    professional,
    on='Metro'
)

college2024 = college2024.merge(
    doctorate,
    on='Metro'
)

In [267]:
college2024['CollegeEducated2024'] = (
    college2024['Bachelors']
    + college2024['Masters']
    + college2024['Professional']
    + college2024['Doctorate']
)

In [268]:
college2024 = college2024[
    ['Metro', 'CollegeEducated2024']
]

In [269]:
college2024

,Metro,CollegeEducated2024
0,"Abilene, TX",30447
1,"Aguadilla, PR",46302
2,"Akron, OH",172473
3,"Albany, GA",22435
4,"Albany, OR",19848
...,...,...
388,"Yakima, WA",30258
389,"York-Hanover, PA",89837
390,"Youngstown-Warren, OH",73898
391,"Yuba City, CA",23177


In [270]:
college2024.to_csv(
    'Data Intermediate/college2024_tidy.csv',
    index=False
)

In [271]:
college2024_for_merge = (
    college2024
    .rename(columns={'Metro': 'CensusMetro'})
)

In [272]:
college2024_mapped = (
    top100_zillow_mapped.merge(
        college2024_for_merge,
        on='CensusMetro',
        how='left',
        validate='one_to_one'
    )
)

In [273]:
print(
    "College matches:",
    college2024_mapped['CollegeEducated2024'].notna().sum()
)

College matches: 65


In [274]:
income_population_metros = set(
    analysis_2024_top100.loc[
        analysis_2024_top100['Income2024'].notna(),
        'Metro'
    ]
)

college_metros = set(
    college2024_mapped.loc[
        college2024_mapped['CollegeEducated2024'].notna(),
        'Metro'
    ]
)

income_population_metros - college_metros

{'Modesto, CA'}

In [275]:
analysis_2024_top100[
    analysis_2024_top100['Metro'].isin(
        income_population_metros - college_metros
    )
][['Metro', 'CensusMetro']]

,Metro,CensusMetro
99,"Modesto, CA","Modesto, CA"


In [277]:
college2024[
    college2024['Metro'].str.contains(
        'Modesto, CA',
        case=False,
        na=False
    )
]

,Metro,CollegeEducated2024
238,"Modesto, CA",69633


In [278]:
analysis_2024_top100 = analysis_2024_top100.merge(
    college2024_mapped[
        ['Metro', 'CollegeEducated2024']
    ],
    on='Metro',
    how='left',
    validate='one_to_one'
)

In [279]:
analysis_2024_top100[
    [
        'Income2024',
        'Population2024',
        'CollegeEducated2024'
    ]
].notna().sum()

Income2024             66
Population2024         66
CollegeEducated2024    65
dtype: int64

In [280]:
analysis_2024_top100['CollegeEducatedPct'] = (
    analysis_2024_top100['CollegeEducated2024']
    / analysis_2024_top100['Population2024']
)

In [281]:
analysis_2024_top100

,RegionID,SizeRank,Metro,RegionType,StateName,Date,HomeValue,Year,CensusMetro,Income2024,Population2024,CollegeEducated2024,CollegeEducatedPct
0,394913,1,"New York, NY",msa,NY,2024-12-31,683398.809012,2024,"New York-Newark-Jersey City, NY-NJ",99852.0,19798537.0,6168034.0,0.311540
1,753899,2,"Los Angeles, CA",msa,CA,2024-12-31,980928.090661,2024,"Los Angeles-Long Beach-Anaheim, CA",96405.0,12974487.0,3468765.0,0.267353
2,394463,3,"Chicago, IL",msa,IL,2024-12-31,331370.219996,2024,"Chicago-Naperville-Elgin, IL-IN",90770.0,9371595.0,2696647.0,0.287747
3,394514,4,"Dallas, TX",msa,TX,2024-12-31,385136.186667,2024,"Dallas-Fort Worth-Arlington, TX",92733.0,7985590.0,2067948.0,0.258960
4,394692,5,"Houston, TX",msa,TX,2024-12-31,316340.099352,2024,"Houston-Pasadena-The Woodlands, TX",81417.0,7442788.0,1742444.0,0.234112
...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,394666,99,"Harrisburg, PA",msa,PA,2024-12-31,295320.038930,2024,"Harrisburg, PA",NaN,NaN,NaN,NaN
96,395113,100,"Spokane, WA",msa,WA,2024-12-31,421330.521046,2024,"Spokane, WA",NaN,NaN,NaN,NaN
97,395075,101,"Scranton, PA",msa,PA,2024-12-31,214263.905867,2024,"Scranton, PA",NaN,NaN,NaN,NaN
98,394460,102,"Chattanooga, TN",msa,TN,2024-12-31,322983.871174,2024,"Chattanooga, TN",NaN,NaN,NaN,NaN


In [282]:
analysis_2024_top100.to_csv(
    'Data Intermediate/analysis_2024_top100.csv',
    index=False
)

In [283]:
analysis_2024_top100.shape

(100, 13)

In [284]:
analysis_2024_top100_check = pd.read_csv(
    'Data Intermediate/analysis_2024_top100.csv'
)

analysis_2024_top100_check.head()

,RegionID,SizeRank,Metro,RegionType,StateName,Date,HomeValue,Year,CensusMetro,Income2024,Population2024,CollegeEducated2024,CollegeEducatedPct
0,394913,1,"New York, NY",msa,NY,2024-12-31,683398.809012,2024,"New York-Newark-Jersey City, NY-NJ",99852.0,19798537.0,6168034.0,0.311540
1,753899,2,"Los Angeles, CA",msa,CA,2024-12-31,980928.090661,2024,"Los Angeles-Long Beach-Anaheim, CA",96405.0,12974487.0,3468765.0,0.267353
2,394463,3,"Chicago, IL",msa,IL,2024-12-31,331370.219996,2024,"Chicago-Naperville-Elgin, IL-IN",90770.0,9371595.0,2696647.0,0.287747
3,394514,4,"Dallas, TX",msa,TX,2024-12-31,385136.186667,2024,"Dallas-Fort Worth-Arlington, TX",92733.0,7985590.0,2067948.0,0.258960
4,394692,5,"Houston, TX",msa,TX,2024-12-31,316340.099352,2024,"Houston-Pasadena-The Woodlands, TX",81417.0,7442788.0,1742444.0,0.234112


In [285]:
analysis_2024_top100.columns

Index(['RegionID', 'SizeRank', 'Metro', 'RegionType', 'StateName', 'Date',
       'HomeValue', 'Year', 'CensusMetro', 'Income2024', 'Population2024',
       'CollegeEducated2024', 'CollegeEducatedPct'],
      dtype='object')

## Data Dictionary

| Column | Description |
|---------|-------------|
| **RegionID** | Unique Zillow identifier for each metropolitan region. |
| **SizeRank** | Zillow ranking of metropolitan areas by population size. Lower values indicate larger metros. |
| **Metro** | Zillow metropolitan area name. |
| **RegionType** | Geographic region classification provided by Zillow (e.g., Metro). |
| **StateName** | State in which the metropolitan area is located. |
| **Date** | Observation date (December 2024 snapshot). |
| **HomeValue** | Zillow Home Value Index (ZHVI), representing the typical home value within the metropolitan area. |
| **Year** | Calendar year corresponding to the observation. |
| **CensusMetro** | Standardized Census metropolitan area name used to merge ACS datasets with Zillow data. |
| **Income2024** | Median household income (2024 American Community Survey). |
| **Population2024** | Total metropolitan population (2024 American Community Survey). |
| **CollegeEducated2024** | Number of residents age 25 and older with a bachelor's degree or higher (2024 American Community Survey). |
| **CollegeEducatedPct** | Proportion of residents with a bachelor's degree or higher. Calculated as `CollegeEducated2024 / Population2024` in the current analysis. |

In [286]:
unmatched = analysis_2024_top100[
    analysis_2024_top100['Income2024'].isna()
][['Metro', 'CensusMetro']]

unmatched

,Metro,CensusMetro
41,"Raleigh, NC","Raleigh, NC"
42,"Memphis, TN","Memphis, TN"
44,"Louisville, KY","Louisville, KY"
45,"New Orleans, LA","New Orleans, LA"
46,"Salt Lake City, UT","Salt Lake City, UT"
47,"Hartford, CT","Hartford, CT"
48,"Buffalo, NY","Buffalo, NY"
51,"Grand Rapids, MI","Grand Rapids, MI"
57,"Omaha, NE","Omaha, NE"
58,"Bridgeport, CT","Bridgeport, CT"


In [287]:
missing_mapping = top100_zillow[
    ~top100_zillow['Metro'].isin(metro_mapping['ZillowMetro'])
][['Metro']]

missing_mapping

,Metro
267632,"Pittsburgh, PA"
267637,"Columbus, OH"
267639,"Cleveland, OH"
267644,"Jacksonville, FL"
267646,"Oklahoma City, OK"
...,...
267698,"Augusta, GA"
267699,"Palm Bay, FL"
267700,"Jackson, MS"
267701,"Harrisburg, PA"


In [288]:
unmatched['Metro'].tolist()

['Raleigh, NC',
 'Memphis, TN',
 'Louisville, KY',
 'New Orleans, LA',
 'Salt Lake City, UT',
 'Hartford, CT',
 'Buffalo, NY',
 'Grand Rapids, MI',
 'Omaha, NE',
 'Bridgeport, CT',
 'Greenville, SC',
 'Bakersfield, CA',
 'Albany, NY',
 'McAllen, TX',
 'Allentown, PA',
 'Oxnard, CA',
 'North Port, FL',
 'Charleston, SC',
 'Greensboro, NC',
 'Stockton, CA',
 'Cape Coral, FL',
 'Little Rock, AR',
 'Lakeland, FL',
 'Des Moines, IA',
 'Winston, NC',
 'Deltona, FL',
 'Provo, UT',
 'Durham, NC',
 'Augusta, GA',
 'Palm Bay, FL',
 'Harrisburg, PA',
 'Spokane, WA',
 'Scranton, PA',
 'Chattanooga, TN']

In [290]:
for metro in unmatched['Metro']:
    city = metro.split(',')[0]

    matches = median_income2024_tidy[
        median_income2024_tidy['Metro'].str.contains(
            city,
            case=False,
            na=False
        )
    ]['Metro'].tolist()

    print(f"\n{metro}")
    print(matches)


Raleigh, NC
['Raleigh-Cary, NC']

Memphis, TN
['Memphis, TN-MS-AR']

Louisville, KY
['Louisville/Jefferson County, KY-IN']

New Orleans, LA
['New Orleans-Metairie, LA']

Salt Lake City, UT
['Salt Lake City-Murray, UT']

Hartford, CT
['Hartford-West Hartford-East Hartford, CT']

Buffalo, NY
['Buffalo-Cheektowaga, NY']

Grand Rapids, MI
['Grand Rapids-Wyoming-Kentwood, MI']

Omaha, NE
['Omaha, NE-IA']

Bridgeport, CT
['Bridgeport-Stamford-Danbury, CT']

Greenville, SC
['Greenville, NC', 'Greenville-Anderson-Greer, SC']

Bakersfield, CA
['Bakersfield-Delano, CA']

Albany, NY
['Albany, GA', 'Albany, OR', 'Albany-Schenectady-Troy, NY']

McAllen, TX
['McAllen-Edinburg-Mission, TX']

Allentown, PA
['Allentown-Bethlehem-Easton, PA-NJ']

Oxnard, CA
['Oxnard-Thousand Oaks-Ventura, CA']

North Port, FL
['North Port-Bradenton-Sarasota, FL']

Charleston, SC
['Charleston, WV', 'Charleston-North Charleston, SC']

Greensboro, NC
['Greensboro-High Point, NC']

Stockton, CA
['Stockton-Lodi, CA']

Cape 

In [292]:
new_mappings = pd.DataFrame({
    'ZillowMetro': [
        'Raleigh, NC',
        'Memphis, TN',
        'Louisville, KY',
        'New Orleans, LA',
        'Salt Lake City, UT',
        'Hartford, CT',
        'Buffalo, NY',
        'Grand Rapids, MI',
        'Omaha, NE',
        'Bridgeport, CT',
        'Greenville, SC',
        'Bakersfield, CA',
        'Albany, NY',
        'McAllen, TX',
        'Allentown, PA',
        'Oxnard, CA',
        'North Port, FL',
        'Charleston, SC',
        'Greensboro, NC',
        'Stockton, CA',
        'Cape Coral, FL',
        'Little Rock, AR',
        'Lakeland, FL',
        'Des Moines, IA',
        'Winston, NC',
        'Deltona, FL',
        'Provo, UT',
        'Durham, NC',
        'Augusta, GA',
        'Palm Bay, FL',
        'Harrisburg, PA',
        'Spokane, WA',
        'Scranton, PA',
        'Chattanooga, TN'
    ],
    'CensusMetro': [
        'Raleigh-Cary, NC',
        'Memphis, TN-MS-AR',
        'Louisville/Jefferson County, KY-IN',
        'New Orleans-Metairie, LA',
        'Salt Lake City-Murray, UT',
        'Hartford-West Hartford-East Hartford, CT',
        'Buffalo-Cheektowaga, NY',
        'Grand Rapids-Wyoming-Kentwood, MI',
        'Omaha, NE-IA',
        'Bridgeport-Stamford-Danbury, CT',
        'Greenville-Anderson-Greer, SC',
        'Bakersfield-Delano, CA',
        'Albany-Schenectady-Troy, NY',
        'McAllen-Edinburg-Mission, TX',
        'Allentown-Bethlehem-Easton, PA-NJ',
        'Oxnard-Thousand Oaks-Ventura, CA',
        'North Port-Bradenton-Sarasota, FL',
        'Charleston-North Charleston, SC',
        'Greensboro-High Point, NC',
        'Stockton-Lodi, CA',
        'Cape Coral-Fort Myers, FL',
        'Little Rock-North Little Rock-Conway, AR',
        'Lakeland-Winter Haven, FL',
        'Des Moines-West Des Moines, IA',
        'Winston-Salem, NC',
        'Deltona-Daytona Beach-Ormond Beach, FL',
        'Provo-Orem-Lehi, UT',
        'Durham-Chapel Hill, NC',
        'Augusta-Richmond County, GA-SC',
        'Palm Bay-Melbourne-Titusville, FL',
        'Harrisburg-Carlisle, PA',
        'Spokane-Spokane Valley, WA',
        'Scranton--Wilkes-Barre, PA',
        'Chattanooga, TN-GA'
    ]
})

In [294]:
new_mappings

,ZillowMetro,CensusMetro
0,"Raleigh, NC","Raleigh-Cary, NC"
1,"Memphis, TN","Memphis, TN-MS-AR"
2,"Louisville, KY","Louisville/Jefferson County, KY-IN"
3,"New Orleans, LA","New Orleans-Metairie, LA"
4,"Salt Lake City, UT","Salt Lake City-Murray, UT"
5,"Hartford, CT","Hartford-West Hartford-East Hartford, CT"
6,"Buffalo, NY","Buffalo-Cheektowaga, NY"
7,"Grand Rapids, MI","Grand Rapids-Wyoming-Kentwood, MI"
8,"Omaha, NE","Omaha, NE-IA"
9,"Bridgeport, CT","Bridgeport-Stamford-Danbury, CT"


In [295]:
metro_mapping = pd.concat(
    [metro_mapping, new_mappings],
    ignore_index=True
)

In [296]:
metro_mapping = metro_mapping.drop_duplicates(
    subset='ZillowMetro',
    keep='last'
)

In [297]:
metro_mapping.to_csv(
    'Data Intermediate/metro_mapping.csv',
    index=False
)

In [298]:
metro_mapping

,ZillowMetro,CensusMetro
0,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"New York, NY","New York-Newark-Jersey City, NY-NJ"
3,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
4,"Atlanta, GA","Atlanta-Sandy Springs-Roswell, GA"
...,...,...
65,"Palm Bay, FL","Palm Bay-Melbourne-Titusville, FL"
66,"Harrisburg, PA","Harrisburg-Carlisle, PA"
67,"Spokane, WA","Spokane-Spokane Valley, WA"
68,"Scranton, PA","Scranton--Wilkes-Barre, PA"


In [299]:
analysis_2024_top100 = top100_zillow.copy()

In [300]:
analysis_2024_top100.shape

(97, 4)

In [301]:
top100_zillow.shape

(97, 4)

In [302]:
zillow_2024.shape

(894, 8)

In [303]:
zillow_2024.columns

Index(['RegionID', 'SizeRank', 'Metro', 'RegionType', 'StateName', 'Date',
       'HomeValue', 'Year'],
      dtype='object')

In [304]:
top100_zillow = (
    zillow_2024
    .sort_values('SizeRank')
    .drop_duplicates(subset='Metro')
    .head(100)
    .copy()
)

In [305]:
top100_zillow.shape

(100, 8)

In [306]:
top100_zillow['Metro'].nunique()

100

In [307]:
analysis_2024_top100 = top100_zillow.copy()

In [308]:
analysis_2024_top100 = analysis_2024_top100.merge(
    metro_mapping,
    left_on='Metro',
    right_on='ZillowMetro',
    how='left'
)

In [309]:
analysis_2024_top100['CensusMetro'] = (
    analysis_2024_top100['CensusMetro']
    .fillna(analysis_2024_top100['Metro'])
)

In [310]:
analysis_2024_top100 = analysis_2024_top100.drop(
    columns='ZillowMetro'
)

In [311]:
print(analysis_2024_top100.shape)
analysis_2024_top100[['Metro', 'CensusMetro']].head()

(100, 9)


,Metro,CensusMetro
0,"New York, NY","New York-Newark-Jersey City, NY-NJ"
1,"Los Angeles, CA","Los Angeles-Long Beach-Anaheim, CA"
2,"Chicago, IL","Chicago-Naperville-Elgin, IL-IN"
3,"Dallas, TX","Dallas-Fort Worth-Arlington, TX"
4,"Houston, TX","Houston-Pasadena-The Woodlands, TX"


In [312]:
analysis_2024_top100 = analysis_2024_top100.merge(
    median_income2024_tidy,
    left_on='CensusMetro',
    right_on='Metro',
    how='left',
    suffixes=('', '_income')
)

In [313]:
analysis_2024_top100 = analysis_2024_top100.drop(
    columns='Metro_income',
    errors='ignore'
)

In [314]:
print('Shape:', analysis_2024_top100.shape)
print(
    'Income matches:',
    analysis_2024_top100['Income2024'].notna().sum()
)

Shape: (100, 10)
Income matches: 100


In [318]:
analysis_2024_top100 = analysis_2024_top100.merge(
    population2024_tidy,
    on='CensusMetro',
    how='left'
)

In [319]:
print(analysis_2024_top100.shape)

print(
    'Population matches:',
    analysis_2024_top100['Population2024'].notna().sum()
)

(100, 11)
Population matches: 100


In [320]:
college2024.columns

Index(['Metro', 'CollegeEducated2024'], dtype='object')

In [321]:
analysis_2024_top100 = analysis_2024_top100.merge(
    college2024,
    left_on='CensusMetro',
    right_on='Metro',
    how='left',
    suffixes=('', '_college')
)

In [322]:
analysis_2024_top100 = analysis_2024_top100.drop(
    columns='Metro_college',
    errors='ignore'
)

In [323]:
print(analysis_2024_top100.shape)

print(
    'College matches:',
    analysis_2024_top100['CollegeEducated2024'].notna().sum()
)

(100, 12)
College matches: 100


In [324]:
print(
    "Income matches:",
    analysis_2024_top100["Income2024"].notna().sum()
)

Income matches: 100


In [325]:
analysis_2024_top100["CollegeEducatedPct"] = (
    analysis_2024_top100["CollegeEducated2024"]
    / analysis_2024_top100["Population2024"]
)

In [326]:
analysis_2024_top100["PriceToIncomeRatio"] = (
    analysis_2024_top100["HomeValue"]
    / analysis_2024_top100["Income2024"]
)

In [327]:
print("Final shape:", analysis_2024_top100.shape)

print(
    "Duplicate metros:",
    analysis_2024_top100["Metro"].duplicated().sum()
)

print(
    "Missing values:"
)

print(
    analysis_2024_top100[
        [
            "Income2024",
            "Population2024",
            "CollegeEducated2024",
            "CollegeEducatedPct",
            "PriceToIncomeRatio"
        ]
    ].isna().sum()
)

Final shape: (100, 14)
Duplicate metros: 0
Missing values:
Income2024             0
Population2024         0
CollegeEducated2024    0
CollegeEducatedPct     0
PriceToIncomeRatio     0
dtype: int64


In [328]:
analysis_2024_top100[
    [
        "Metro",
        "HomeValue",
        "Income2024",
        "Population2024",
        "CollegeEducated2024",
        "CollegeEducatedPct",
        "PriceToIncomeRatio"
    ]
].head()

,Metro,HomeValue,Income2024,Population2024,CollegeEducated2024,CollegeEducatedPct,PriceToIncomeRatio
0,"New York, NY",683398.809012,99852.0,19798537.0,6168034,0.311540,6.844117
1,"Los Angeles, CA",980928.090661,96405.0,12974487.0,3468765,0.267353,10.175075
2,"Chicago, IL",331370.219996,90770.0,9371595.0,2696647,0.287747,3.650658
3,"Dallas, TX",385136.186667,92733.0,7985590.0,2067948,0.258960,4.153173
4,"Houston, TX",316340.099352,81417.0,7442788.0,1742444,0.234112,3.885431


In [329]:
analysis_2024_top100.to_csv(
    "Data Intermediate/analysis_2024_top100.csv",
    index=False
)

In [330]:
analysis_2024_check = pd.read_csv(
    "Data Intermediate/analysis_2024_top100.csv"
)

print(analysis_2024_check.shape)
analysis_2024_check.columns

(100, 14)


Index(['RegionID', 'SizeRank', 'Metro', 'RegionType', 'StateName', 'Date',
       'HomeValue', 'Year', 'CensusMetro', 'Income2024', 'Population2024',
       'CollegeEducated2024', 'CollegeEducatedPct', 'PriceToIncomeRatio'],
      dtype='object')